---
title: "Getting Started with STdeconvolve"
author: "Jean Fan"
date: "2/26/2021"
output:   
  md_document:
    variant: markdown_github
vignette: >
  %\VignetteIndexEntry{Getting Started with STdeconvolve}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load package

```{r, getting_started_package}
library(STdeconvolve)
```

Load built in data

```{r, data}
data(mOB)
pos <- mOB$pos
cd <- mOB$counts
annot <- mOB$annot
```

Feature select

```{r, getting_started_feature, fig.width=6, fig.height=3}
## remove pixels with too few genes
counts <- cleanCounts(cd)
## feature select for genes
corpus <- restrictCorpus(counts, t1=0.05, t2 = 1)
```

Choose optimal number of cell-types

```{r, getting_started_opt, fig.width=6, fig.height=4}
lda <- fitLDA(corpus, Ks = seq(2, 9, by = 1), plot=TRUE, verbose=FALSE)
```

Get best model results

```{r, getting_started_model}
results <- getBetaTheta(lda$models[[lda$kOpt2]])
deconProp <- results$theta
deconGexp <- results$beta*1000
```

Visualize deconvolved cell-type proportions

```{r, getting_started_proportions, fig.width=8, fig.height=4}
vizAllTopics(deconProp, pos, 
             groups = annot, 
             group_cols = rainbow(length(levels(annot))),
             r=0.4)
```

Visualize deconvolved cell-type gene expression

```{r, getting_started_expression, fig.width=6, fig.height=4}
celltype <- 7
highgexp <- names(which(deconGexp[celltype,] > 5))
log2fc <- sort(log2(deconGexp[celltype,highgexp]/colMeans(deconGexp[-celltype,highgexp])), decreasing=TRUE)
markers <- names(log2fc)[1:4]

df <- merge(as.data.frame(pos), 
            as.data.frame(t(as.matrix(counts[markers,]))), 
            by = 0)
ps <- lapply(markers, function(marker) {
  vizGeneCounts(df = df,
              gene = marker,
              size = 2, stroke = 0.1,
              plotTitle = marker,
              winsorize = 0.05)
})
gridExtra::grid.arrange(
  grobs = ps,
  layout_matrix = rbind(c(1, 2),
                        c(3, 4))
)

celltype <- 2
highgexp <- names(which(deconGexp[celltype,] > 5))
log2fc <- sort(log2(deconGexp[celltype,highgexp]/colMeans(deconGexp[-celltype,highgexp])), decreasing=TRUE)
markers <- names(log2fc)[1:4]

df <- merge(as.data.frame(pos), 
            as.data.frame(t(as.matrix(counts[markers,]))), 
            by = 0)
ps <- lapply(markers, function(marker) {
  vizGeneCounts(df = df,
              gene = marker,
              size = 2, stroke = 0.1,
              plotTitle = marker,
              winsorize = 0.05)
})
gridExtra::grid.arrange(
  grobs = ps,
  layout_matrix = rbind(c(1, 2),
                        c(3, 4))
)
```



