---
title: "pdac_moncada_gse111672"
author: "Brendan F. Miller"
date: "2/4/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Moncada et al GSE111672

ST and scRNA-seq data so have a reference for cell types


# data

get filtered St data for PDAC-A ST1. Apparently it is filtered already?

Overlay on jpeg image. Make sure everything looks logical

```{r}

pdacAst1Path <- "/Users/brendan/Desktop/PostDoc/work/STDeconvolve/data/moncada_pdac_GSE111672/GSE111672_RAW/GSM3036911_PDAC-A-ST1-filtered.txt"
pdacAst1 <- read.table(pdacAst1Path, header = TRUE, sep = "\t")
pdacAst1

```

coordinates of spots:
```{r}

spotIDs <- unlist(lapply(colnames(pdacAst1)[2:429], function(i) {
  ix <- strsplit(i, "X")[[1]][2]
  ix
}))

pdacAst1.Pos <- do.call(rbind, lapply(spotIDs, function(ix) {
  coords <- as.numeric(strsplit(ix, "x")[[1]])
  coords
}))

colnames(pdacAst1.Pos) <- c("x", "y")
rownames(pdacAst1.Pos) <- spotIDs

```

reformat count mtx:
```{r}

pdacAst1.t <- t(pdacAst1)
colnames(pdacAst1.t) <- pdacAst1$Genes
pdacAst1.t <- pdacAst1.t[2:429,]
pdacAst1 <- apply(pdacAst1.t, 2,FUN = as.numeric)
rownames(pdacAst1) <- rownames(pdacAst1.Pos)

```

visualize spots:
```{r}

# imagepath <- "/Users/brendan/Desktop/PostDoc/work/STDeconvolve/data/moncada_pdac_GSE111672/GSE111672_RAW/GSM3036911_PDAC-A-ST1-HE.jpg"
# 
# jpg <- readJPEG(imagepath)
# dim(jpg)

plot.new()
plot(pdacAst1.Pos[,"x"], pdacAst1.Pos[,"y"], col='red', pch="+")
# 
# rasterImage(jpg[dim(jpg)[1]:1, , ],
#             0, 0,
#             dim(jpg)[2], dim(jpg)[1])
# points(pdacAst1.Pos[,"x"], pdacAst1.Pos[,"y"], col='red', pch="+")

```
check out the gene counts
```{r, mob-qc, fig.width=8, fig.height=3}

# Remove poor datasets and genes
pdacAst1.Clean <- MERINGUE::cleanCounts(counts = t(pdacAst1), # genes x spots mtx
                      min.reads = 100, 
                      min.lib.size = 100, 
                      plot=TRUE,
                      verbose=TRUE)

# CPM normalize
pdacAst1.CPM <- MERINGUE::normalizeCounts(counts = pdacAst1.Clean, 
                       log=FALSE,
                       verbose=TRUE)

```

```{r}

# using the overdispersed genes (derived from the raw counts) and the values of the the CPM count matrix

# Dimensionality reduction by PCA on log10 CPM expression values
pcs.info <- prcomp(t(log10(as.matrix(pdacAst1.CPM)+1)), center=TRUE)
nPcs <- 5
pcs <- pcs.info$x[,1:nPcs]

# 2D embedding by tSNE
emb <- Rtsne::Rtsne(pcs,
             is_distance=FALSE,
             perplexity=30,
             num_threads=1,
             verbose=FALSE)$Y
rownames(emb) <- rownames(pcs)

# Graph-based cluster detection
k <- 30
com <- getClusters(pcs, k, weight=TRUE)

# names(com) <- rownames(pdacAst1.Pos)

# Manually annotate identified clusters with cell-types
# annot <- as.character(com); names(annot) <- names(com)
# annot[com==4] <- '1: Granule Cell Layer'
# annot[com==1] <- '2: Mitral Cell Layer'
# annot[com==3] <- '3: Outer Plexiform Layer'
# annot[com==2] <- '4: Glomerular Layer'
# annot[com==5] <- '5: Olfactory Nerve Layer'
# annot <- as.factor(annot)

# Plot
par(mfrow=c(1,2), mar=rep(1,4))
# plotEmbedding(emb, groups=annot, 
#               show.legend=TRUE, xlab=NA, ylab=NA,
#               verbose=FALSE)
# 
# plotEmbedding(posMob, groups=annot, 
#               cex=1, xlab=NA, ylab=NA,
#               verbose=FALSE)

plotEmbedding(emb, groups=com, 
              show.legend=TRUE, xlab=NA, ylab=NA,
              verbose=FALSE)

plotEmbedding(pdacAst1.Pos, groups=com,
              cex=1, xlab=NA, ylab=NA,
              verbose=FALSE)

```
one detail: MUSE transcripts showed 4 clusters. Maybe try clustering on most variable genes like they did. Will have to do this anyways for lda

```{r}

top_expressed <- names(rowSums(pdacAst1.Clean)[order(rowSums(pdacAst1.Clean),
                                             decreasing = TRUE)][1:5])
# top_expressed
# "REG3A"  "CTRB2"  "PRSS2"  "CTRB1"  "S100A6"
# 20739  14839  10682  10317   8709 
# actually, these genes could be relevant. Keep for now.

dim(pdacAst1.Clean)
# instead, remove ribosomal and mitochondrial genes:
pdacAst1.Filt <- pdacAst1.Clean[!grepl("^MT", rownames(pdacAst1.Clean)),]
dim(pdacAst1.Filt)
pdacAst1.Filt <- pdacAst1.Filt[!grepl("^RPL", rownames(pdacAst1.Filt)),]
dim(pdacAst1.Filt)
pdacAst1.Filt <- pdacAst1.Filt[!grepl("^MRPL", rownames(pdacAst1.Filt)),]
dim(pdacAst1.Filt)

```

Remove words that appear in every document

```{r}

# remove genes that are present in all spots
countsFilt_ <- pdacAst1.Filt
countsFilt_[which(countsFilt_ > 0)] <- 1
pdacAst1.Filt <- pdacAst1.Filt[which(rowSums(countsFilt_) < 260),]
dim(pdacAst1.Filt)

```

```{r, fig.width=6, fig.height=3}

# $mat -> normalized matrix for all genes
# $ods -> list of the overdispersed genes
# $df -> table of stats for all the input genes

par(mfrow=c(4,2), mar=c(1,1,1,1))

pdacAst1odGenes <- getOverdispersedGenes(pdacAst1.Filt,
                                    plot = TRUE,
                                    details = TRUE)

```

```{r}

pdacAst1.sigGenes <- pdacAst1.Filt[pdacAst1odGenes$ods,]

top_expressed <- rowSums(pdacAst1.sigGenes)[order(rowSums(pdacAst1.sigGenes),
                                             decreasing = TRUE)][1:5]
top_expressed

```


439 OD genes. Cluster with these and use raw counts

```{r}

# using the overdispersed genes (derived from the CPM counts) and the values of the the CPM count matrix

pcs.info <- prcomp(t(log10(as.matrix(pdacAst1.sigGenes)+1)), center=TRUE)
nPcs <- 5
pcs <- pcs.info$x[,1:nPcs]

# 2D embedding by tSNE
emb2 <- Rtsne::Rtsne(pcs,
             is_distance=FALSE,
             perplexity=30,
             num_threads=1,
             verbose=FALSE)$Y
rownames(emb2) <- rownames(pcs)

# Graph-based cluster detection
k <- 30
com2 <- getClusters(pcs, k, weight=TRUE)

par(mfrow=c(1,2), mar=rep(1,4))

plotEmbedding(emb2, groups=com2, 
              show.legend=TRUE, xlab=NA, ylab=NA,
              verbose=FALSE)
plotEmbedding(pdacAst1.Pos, groups=com2, 
              cex=1, xlab=NA, ylab=NA,
              verbose=FALSE)

# from before:
plotEmbedding(emb, groups=com, 
              show.legend=TRUE, xlab=NA, ylab=NA,
              verbose=FALSE)

plotEmbedding(pdacAst1.Pos, groups=com,
              cex=1, xlab=NA, ylab=NA,
              verbose=FALSE)

```
the txn clusters on the tissue look somewhat similar but a few differences.

## corpus

```{r}

# docs x terms

pdac_corpus <- t(as.matrix(pdacAst1.sigGenes))
pdac_corpus_slm <- slam::as.simple_triplet_matrix(pdac_corpus)

```

# fit

```{r}

k_ <- seq(from = 2, to = 10, by = 1)
k_[1] <- 2
k_ <- append(k_, seq(from = 15, to = 50, by = 5))
# [1]  2  3  4  5  6  7  8  9 10 15 20 25 30 35 40 45 50

```

## opt

```{r}

pdac_lda <- fitLDA(pdac_corpus_slm, k_)

```

```{r}

pdac_lda.tmResult <- posterior(pdac_lda)
pdac_lda.theta <- pdac_lda.tmResult$topics
pdac_lda.beta <- pdac_lda.tmResult$terms
pdac_lda.topicFreqsOverall <- colSums(pdac_lda.theta) / nrow(pdac_corpus_slm)

```

## excess

```{r}

k_excess <- seq(from = 50, to = 200, by = 50)
# k_[1] <- 2
# k_ <- append(k_, seq(from = 15, to = 50, by = 5))
# [1]  2  3  4  5  6  7  8  9 10 15 20 25 30 35 40 45 50

```

```{r}

pdac_lda_excess <- fitLDA(pdac_corpus_slm, k_excess)

```

actually this is better than 50 apparantly
```{r}

pdac_lda_excess
pdac.k150 <- pdac_lda_excess

```

```{r}

pdac.k150.tmResult <- posterior(pdac.k150)
pdac.k150.theta <- pdac.k150.tmResult$topics
pdac.k150.beta <- pdac.k150.tmResult$terms
pdac.k150.topicFreqsOverall <- colSums(pdac.k150.theta) / nrow(pdac_corpus_slm)

```

# clusters

```{r}

pdac_lda.clust <- clusterTopics(beta = pdac_lda.beta,
                               distance = "euclidean",
                               clustering = "ward.D",
                               dynamic = "hybrid",
                               deepSplit = 4,
                               plotDendro = TRUE)

pdac.k150.clust <- clusterTopics(beta = pdac.k150.beta,
                               distance = "euclidean",
                               clustering = "ward.D",
                               dynamic = "hybrid",
                               deepSplit = 4,
                               plotDendro = TRUE)

```

```{r}

clusterColorsk50 <- pdac_lda.clust$clusters
levels(clusterColorsk50) <- gg_color_hue(length(levels(clusterColorsk50)))

clusterColorsk150 <- pdac.k150.clust$clusters
levels(clusterColorsk150) <- gg_color_hue(length(levels(clusterColorsk150)))

```

# ---------------------------------------------
# Correlation - beta

## proxy beta

use "com", which is clustering on CPM normalized all genes after filtering

```{r}

betaProxy <- do.call(rbind, lapply(levels(com), function(cell_layer){
  
  layerCounts <- colSums(pdac_corpus[names(com[which(com == cell_layer)]),])
  layerProportions <- layerCounts / sum(layerCounts)
  layerProportions
  
}))

rownames(betaProxy) <- levels(com)

```

```{r}

pdac_lda.betaCombined <- combineTopics(mtx = pdac_lda.beta, clusters = clusterColorsk50, mtxType = "b")

pdac.k150.betaCombined <- combineTopics(mtx = pdac.k150.beta, clusters = clusterColorsk150, mtxType = "b")

```

## -- all topics

```{r}

pdac_lda.betaGTcorMtx <- correlationBetweenBetas(beta1 = pdac_lda.beta,
                                                beta2 = betaProxy,
                                                thresh = NULL)

pdac.k150.betaGTcorMtx <- correlationBetweenBetas(beta1 = pdac.k150.beta,
                                                beta2 = betaProxy,
                                                thresh = NULL)

pdac_lda.betaCombGTcorMtx <- correlationBetweenBetas(beta1 = pdac_lda.betaCombined,
                                                beta2 = betaProxy,
                                                thresh = NULL)

pdac.k150.betaCombGTcorMtx <- correlationBetweenBetas(beta1 = pdac.k150.betaCombined,
                                                beta2 = betaProxy,
                                                thresh = NULL)

```

```{r, fig.height=7, fig.width=8}

par(mfrow=c(1,1), mar=c(8,8,3,2))
heatmap.2(pdac_lda.betaGTcorMtx,
          Rowv = pdac_lda.clust$dendro,
          density.info = "none",
          trace = "none",
          RowSideColors = as.vector(clusterColorsk50),
          # ColSideColors = as.vector(classColors),
          col = correlation_palette,
          breaks = correlation_breaks,
          cexRow=0.5,cexCol=0.7,margins=c(6,3),
          main = "pdac_lda.betaGTcorMtx",
          lhei = c(1,5),
          key.xlab = "Correlation",
          key.title = NA)

par(mfrow=c(1,1), mar=c(8,8,3,2))
heatmap.2(pdac.k150.betaGTcorMtx,
          Rowv = pdac.k150.clust$dendro,
          density.info = "none",
          trace = "none",
          RowSideColors = as.vector(clusterColorsk150),
          # ColSideColors = as.vector(classColors),
          col = correlation_palette,
          breaks = correlation_breaks,
          cexRow=0.5,cexCol=0.7,margins=c(6,3),
          main = "pdac.k150.betaGTcorMtx",
          lhei = c(1,5),
          key.xlab = "Correlation",
          key.title = NA)


par(mfrow=c(1,1), mar=c(8,8,3,2))
heatmap.2(pdac_lda.betaCombGTcorMtx,
          # Rowv = pdac_lda.clust$dendro,
          density.info = "none",
          trace = "none",
          RowSideColors = rownames(pdac_lda.betaCombGTcorMtx),
          # ColSideColors = as.vector(classColors),
          col = correlation_palette,
          breaks = correlation_breaks,
          cexRow=0.5,cexCol=0.7,margins=c(6,3),
          main = "pdac_lda.betaCombGTcorMtx",
          lhei = c(1,5),
          key.xlab = "Correlation",
          key.title = NA)

par(mfrow=c(1,1), mar=c(8,8,3,2))
heatmap.2(pdac.k150.betaCombGTcorMtx,
          # Rowv = pdac.k150.clust$dendro,
          density.info = "none",
          trace = "none",
          RowSideColors = rownames(pdac.k150.betaCombGTcorMtx),
          # ColSideColors = as.vector(classColors),
          col = correlation_palette,
          breaks = correlation_breaks,
          cexRow=0.5,cexCol=0.7,margins=c(6,3),
          main = "pdac.k150.betaCombGTcorMtx",
          lhei = c(1,5),
          key.xlab = "Correlation",
          key.title = NA)

```


## proxy theta

Because the resolution is at the spot level, will have each spot 100% of a particular cell layer

```{r}

# annot = factor of cell layer values with spot IDs as the labels
thetaProxy <- model.matrix(~ 0 + com)
rownames(thetaProxy) <- names(com)
colnames(thetaProxy) <- levels(com)

```

# ---------------------------------------------
# viz topics theta

visualize the predicted document-topic proportions

correlation to the 5 txn clusters?
Actually check out individual topics or increase cluster res and see if you get sub txn clusters. As additional proof of concepts like the ependymal cells.

```{r}

pdac_lda.thetaCombined <- t(combineTopics(mtx = t(pdac_lda.theta), clusters = clusterColorsk50, mtxType = "t"))

pdac.k150.thetaCombined <- t(combineTopics(mtx = t(pdac.k150.theta), clusters = clusterColorsk150, mtxType = "t"))

```

```{r, fig.height=6, fig.width=8}

vizAllTopics(theta = pdac_lda.thetaCombined,
             pos = pdacAst1.Pos,
             topicOrder = seq_len(length(colnames(pdac_lda.thetaCombined))),
             cluster_cols = colnames(pdac_lda.thetaCombined),
             groups = NA,
             group_cols = NA, 
             r = 0.4,
             lwd = 0.01,
             plotTitle = "k50 pdac")

for (i in colnames(thetaProxy)) {
  vizAllTopics(theta = pdac_lda.thetaCombined,
             pos = pdacAst1.Pos,
             topicOrder = seq_len(length(colnames(pdac_lda.thetaCombined))),
             cluster_cols = colnames(pdac_lda.thetaCombined),
             groups = as.character(thetaProxy[,i]),
             group_cols = c("0" = "black", "1" = "white"),
             r = 0.4,
             lwd = 0.3,
             plotTitle = paste0("k50 Txn C", i))
}


vizAllTopics(theta = pdac.k150.thetaCombined,
             pos = pdacAst1.Pos,
             topicOrder = seq_len(length(colnames(pdac.k150.thetaCombined))),
             cluster_cols = colnames(pdac.k150.thetaCombined),
             groups = NA,
             group_cols = NA, 
             r = 0.4,
             lwd = 0.01,
             plotTitle = "k150 pdac")

for (i in colnames(thetaProxy)) {
  vizAllTopics(theta = pdac.k150.thetaCombined,
             pos = pdacAst1.Pos,
             topicOrder = seq_len(length(colnames(pdac.k150.thetaCombined))),
             cluster_cols = colnames(pdac.k150.thetaCombined),
             groups = as.character(thetaProxy[,i]),
             group_cols = c("0" = "black", "1" = "white"),
             r = 0.4,
             lwd = 0.3,
             plotTitle = paste0("k150 Txn C", i))
}

```
lost a spot, remake positions
```{r}

pdacPos <- do.call(rbind, lapply(rownames(pdac_lda.theta), function(ix) {
  coords <- as.numeric(strsplit(ix, "x")[[1]])
  coords
}))

colnames(pdacPos) <- c("x", "y")
rownames(pdacPos) <- rownames(pdac_lda.theta)

```


```{r, fig.height=6, fig.width=8}

vizTopicClusters(theta = pdac_lda.theta,
                pos = pdacPos,
                topicOrder = pdac_lda.clust$order,
                clusters = clusterColorsk50,
                sharedCol = FALSE,
                r = 0.4,
                lwd = 0.01)

k50thetaCombinedClusters <- as.factor(colnames(pdac_lda.thetaCombined))
names(k50thetaCombinedClusters) <- colnames(pdac_lda.thetaCombined)
vizTopicClusters(theta = pdac_lda.thetaCombined,
                pos = pdacPos,
                topicOrder = seq_len(length(colnames(pdac_lda.thetaCombined))),
                clusters = k50thetaCombinedClusters,
                sharedCol = TRUE,
                r = 0.4,
                lwd = 0.01)

```

```{r, fig.height=6, fig.width=8}

vizTopicClusters(theta = pdac.k150.theta,
                pos = pdacPos,
                topicOrder = pdac.k150.clust$order,
                clusters = clusterColorsk150,
                sharedCol = FALSE,
                r = 0.4,
                lwd = 0.01)

k150thetaCombinedClusters <- as.factor(colnames(pdac.k150.thetaCombined))
names(k150thetaCombinedClusters) <- colnames(pdac.k150.thetaCombined)
vizTopicClusters(theta = pdac.k150.thetaCombined,
                pos = pdacPos,
                topicOrder = seq_len(length(colnames(pdac.k150.thetaCombined))),
                clusters = k150thetaCombinedClusters,
                sharedCol = TRUE,
                r = 0.4,
                lwd = 0.01)

```

orange topic is everywhere..what is it?
A lot of topics are only in a few spots. Could they be noise?

Green could be cancer? Or something else?

Dark green is also interesting

The pink is represented in just the top right across a lot of spots but small amount

Also consider individual topics..


Consider that maybe the gene counts need to be adjusted or normalized? This data is definitely different than the mob..should I select genes differently? "Count" genes differently like binning (like one of the strategies in the gene expression topic model paper)

also the txn clusters look kind of noisy..try to capture the original clusters in paper...


How does k150 compare to the k50 model?


lastly, the scRNAseq data will be a good metric for different cell types (just not their spatial position, however)




k50
```{r}

head(pdac_lda.beta["23",order(pdac_lda.beta["23",], decreasing = TRUE)], n=10)
head(pdac_lda.beta["49",order(pdac_lda.beta["49",], decreasing = TRUE)], n=10)
head(pdac_lda.betaCombined["#7CAE00",order(pdac_lda.betaCombined["#7CAE00",], decreasing = TRUE)], n=10)
head(pdac_lda.beta["5",order(pdac_lda.beta["5",], decreasing = TRUE)], n=10)
head(pdac_lda.beta["43",order(pdac_lda.beta["43",], decreasing = TRUE)], n=10)

```


k150

```{r}

head(pdac.k150.betaCombined["#8B93FF",order(pdac.k150.betaCombined["#8B93FF",], decreasing = TRUE)], n=10)

```







A really important use case may be for a user to input in some sort of prior to search for known cell types. Here the topics are representing biology, but only b/c I know something about the tissue do I know something about the topics. But what does deriving these topics acomplish? It would be ideal if I could actually identify the topics as cell types. But I'm not sure if I can do that yet. I just know something about what they represent. 

Goal: reference free cell type ID. So therefore I need to be able to infer the cell types from the topics, assuming that they are representing biology. And I think a lot of them are. But not so sure about every topic or cluster.

