---
title: "Figure1_proof_of_concept"
author: "Brendan F. Miller"
date: "3/6/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# MERFISH mPOA

```{r}

load("/Users/brendan/Desktop/PostDoc/work/STDeconvolve/data/mpoa_merfish_clean.RData")
# has `annot.table`, `counts`, and `features` objects

# annot.table: table of the individual cells and data like coordinates, cell types, bregma, animal
# features: has cells and additional features. Also dataset they belong to
# counts: gene counts of cell in annot.table for 130 merfish genes profiled

# unique(features$dataset_name)
#  [1] "170908_FN3_2_M22_M26"    "170917_MN1_1pp_M22_M26" 
#  [3] "170919_FN4_1_M22_M26"    "170921_FN4_2_M22_M26"   
#  [5] "170923_FN4_1pp_M22_M26"  "170925_MN2_1p_M22_M26"  
#  [7] "171021_FN7_2_M22_M26"    "171023_FN7_1_M22_M26"   
#  [9] "171024_MN5_2p_M22_M26"   "171027_MN5_1p_M22_M26"  
# [11] "171104_MN6_2_M22_M26"    "171112_MN8_2pppp"       
# [13] "171114_MN8_M22_M26_2pp"  "171208_MN9_2ppp_M22_M26"
# [15] "171118_BD3_1pp"          "171119_BD5_1pp"         
# [17] "171121_BD5_1"            "171123_BD5_2p"          
# [19] "171124_BD6_2ppp"         "171126_BD6_1pppp"       
# [21] "171128_BD7_1ppp"         "171129_BD7_1ppp_actual" 
# [23] "171201_BD_2ppp"          "171203_BD8_1pp"         
# [25] "171210_BD9_1pp"          "171211_BD9_2ppp_M22_M26"
# [27] "171214_BD10_2_M22_M26"   "171216_BD11_1_M22_M26"  
# [29] "171217_BD11_2_M22_M26"   "171219_BD12_1p_M22_M26" 
# [31] "171221_BD12_2p_M22_M26"  "171223_BD13_2p_M22_M26" 

```

Simulate ST spots for each bregma slice for a given MERFISH experiment data set

171021_FN7_2_M22_M26

```{r}

# select cells that are part of given data set:
selected_cells <- rownames(features)[features$dataset_name %in% c('171021_FN7_2_M22_M26')]

spatial_position_and_class <- annot.table[selected_cells,
                                          c('Centroid_X', 'Centroid_Y', 'Bregma', "Cell_class", "Neuron_cluster_ID")]

# remove rows with NA
spatial_position_and_class <- na.omit(spatial_position_and_class) 

```

```{r}

# reduce set of major cell class labels. Convert OD types to "OD" and
# "Endothelial types to "Endothelial"
spatial_position_and_class[grep(pattern = "OD",
                                x = spatial_position_and_class$Cell_class),]$Cell_class <- "OD"

spatial_position_and_class[grep(pattern = "Endothelial",
                                x = spatial_position_and_class$Cell_class),]$Cell_class <- "Endothelial"

# number of cells in the experiment across bregmas
dim(spatial_position_and_class)
# [1] 36329     5

# gene counts for individual cells
bregmaCellsGexp <- counts[rownames(spatial_position_and_class),]

# remove "Blanks" from data
bregmaCellsGexp <- bregmaCellsGexp[,!grepl("Blank", colnames(bregmaCellsGexp))]

```

```{r}

# to expand to neuronal subtypes, make a secondary vector of "Cell_class" where
# "Excitatory and "Inhibitory" are replaces with the respective "Neuron_cluster_ID"

Cell_class_major <- spatial_position_and_class$Cell_class

Cell_class_with_neuronal <- unlist(lapply(rownames(spatial_position_and_class), function(cell){
  class <- spatial_position_and_class[cell,]$Cell_class
  neuron <- spatial_position_and_class[cell,]$Neuron_cluster_ID
  if (neuron != ""){
    i <- neuron
  } else {
    i <- class
  }
  i
}))

length(Cell_class_with_neuronal)

```

The `buildBregmaCorpus` output contains a "gtCtGenes" ground truth cell type - gene expression matrix, but this is built by combining cells of the same Cell_class in the given bregma (and therefore may also be missing reference for cell types not present).

Let's also make a reference using all the cells across the bregmas.

```{r}

cellTypes <- spatial_position_and_class$Cell_class
cells <- rownames(spatial_position_and_class)

mat <- bregmaCellsGexp[cells,]
mm <- model.matrix(~ 0 + factor(cellTypes))
colnames(mm) <- levels(factor(cellTypes))
gtCtGenes <- t(t(as.matrix(mat)) %*% mm)
# remove "Blanks" from data:
# gtCtGenes <- gtCtGenes[,!grepl("Blank", colnames(gtCtGenes))]
gtCtGenes <- gtCtGenes/rowSums(gtCtGenes)

dim(gtCtGenes)

```

```{r}

# simulate spots for each bregma and contain in a hash table
FN7_2_M22_M26 <- simulateBregmaSpots(spatial_position_and_class,
                                          counts = bregmaCellsGexp,
                                          patch_size = 100)

# now simulate but use all the neural types as well
spatial_position_and_class$Cell_class <- Cell_class_with_neuronal
FN7_2_M22_M26_neuro <- simulateBregmaSpots(spatial_position_and_class,
                                          counts = bregmaCellsGexp,
                                          patch_size = 100)

```

Let's also make a reference for all of the cell types including the neuronal ones

```{r}

cellTypes <- spatial_position_and_class$Cell_class
cells <- rownames(spatial_position_and_class)

mat <- bregmaCellsGexp[cells,]
mm <- model.matrix(~ 0 + factor(cellTypes))
colnames(mm) <- levels(factor(cellTypes))
gtCtNeuronalGenes <- t(t(as.matrix(mat)) %*% mm)
# remove "Blanks" from data:
# gtCtNeuronalGenes <- gtCtNeuronalGenes[,!grepl("Blank", colnames(gtCtNeuronalGenes))]
gtCtNeuronalGenes <- gtCtNeuronalGenes/rowSums(gtCtNeuronalGenes)

dim(gtCtNeuronalGenes)

```

Construct corpuses using the simulations with the 9 major cell classes, or expanded to all of the neuronal subtypes

```{r}

simBregmas <- lapply(keys(FN7_2_M22_M26), function(ix){
  bregmma <- buildBregmaCorpus(hashTable = FN7_2_M22_M26, 
                                bregmaID = ix)
  print(bregma04$sim)
  print(dim(bregmma$gtSpotTopics))
  print(dim(bregmma$gtCtGenes))
  bregmma
})
names(simBregmas) <- keys(FN7_2_M22_M26)

simBregmas_neuro <- lapply(keys(FN7_2_M22_M26_neuro), function(ix){
  bregmma <- buildBregmaCorpus(hashTable = FN7_2_M22_M26_neuro, 
                                bregmaID = ix)
  print(bregma04$sim)
  print(dim(bregmma$gtSpotTopics))
  print(dim(bregmma$gtCtGenes))
  bregmma
})
names(simBregmas_neuro) <- keys(FN7_2_M22_M26_neuro)

```

Recall that the `$gtSpotTopics` are of primary interest to compare the fitted models to for each bregma. The `$gtCtGenes` for these corpuses are based only using the cells that were in the given bregma and in simulated patches.


When comparing to reference-dependent methods, could use a single scRNAseq reference made from all of the individual cells in `bregmaCellsGexp`. Or for each bregma, can make a reference based on the individual cells that were present in the spots. So that the models are restricted to the identical data that the LDA is presented,

```{r}

# "scRNAseq" reference using all the cells:
bregmaFullSeur <- CreateSeuratObject(counts = t(bregmaCellsGexp), project = "bregmaFull scRNAseq Ref")

# add cell class labels to the meta.data for assigning clusters later
bregmaFullSeur[["Major_class"]] <- Cell_class_major
bregmaFullSeur[["All_classes"]] <- Cell_class_with_neuronal

```

```{r}

# if using a "scRNAseq" reference for a given bregma in order to restrict to the same data as LDA:

# vectors of the class labels for the cells in the bregma patches:
breg_Cell_class_major <- simBregmas$`-0.04`$annotDf$Cell_class
breg_Cell_class_with_neuronal <- unlist(lapply(rownames(simBregmas$`-0.04`$annotDf), function(cell){
  class <- simBregmas$`-0.04`$annotDf[cell,]$Cell_class
  neuron <- simBregmas$`-0.04`$annotDf[cell,]$Neuron_cluster_ID
  if (neuron != ""){
    i <- neuron
  } else {
    i <- class
  }
  i
}))

# cells in the bregma patches:
breg_patch_cells <- rownames(simBregmas$`-0.04`$annotDf)

# the gene counts for the bregma patch cells
breg_scRNAseq <- FN7_2_M22_M26[["-0.04"]]$cellGexp[breg_patch_cells,]

# "scRNAseq" reference using all the cells:
bregmaSeur <- CreateSeuratObject(counts = t(breg_scRNAseq), project = "bregma scRNAseq Ref")

# add cell class labels to the meta.data for assigning clusters later
bregmaSeur[["Major_class"]] <- breg_Cell_class_major
bregmaSeur[["All_classes"]] <- breg_Cell_class_with_neuronal

```

# MOB

Using the mOB from MERIGNUE and the 12 Stahl et al mOB replicates (13 total data sets), find a common set of genes to use in the corpus for each replicate so that they can be compared equally

```{r}

data(mOB) # from `MERIGNUE`
cd <- mOB$counts # 15928 genes   262 spots

```

```{r}

# list of either the mob data sets or paths to the data for inputs into `preprocess`

mob_paths <- list()
mob_paths[[1]] <- t(cd) # preprocess input needs to be spot x genes

# paths to Stahl data:
p <- "/Users/brendan/Desktop/PostDoc/data/ImputeTranscriptomeFromHE/2016_Stahl/SpotGeneCountMatrices/"
stahlMobDatPaths <- list.files(path = p, pattern = "*MOB_count_matrix-1.tsv", full.names = FALSE)
# add to the list
i <- 1
while(i < length(stahlMobDatPaths) + 1){
  mob_paths[[i+1]] <- paste0(p, stahlMobDatPaths[i])
  i <- i + 1
}

```

```{r}

mobRepGenes <- lapply(mob_paths, function(p) {
  dat <- preprocess(p,
                   alignFile = NA,
                   extractPos = FALSE,
                   selected.genes = NA,
                   nTopGenes = NA,
                   genes.to.remove = NA,
                   perc.spots = NA,
                   min.reads = 100,
                   min.lib.size = 100,
                   ODgenes = TRUE, # select the overdispersed genes
                   od.genes.alpha = 0.25, # increase due to filtering by intersection of reps
                   gam.k = 5)
  colnames(dat$corpus)
})

commonMobGenes <- Reduce(intersect, mobRepGenes)
length(commonMobGenes)

```

Now construct corpuses for each mob dataset using the common set of mob genes

```{r}

# first do the Stahl data because can get positional information from the spot IDs
mobCorpus <- lapply(mob_paths[2:13], function(p) {
  dat <- preprocess(p,
                   alignFile = NA,
                   extractPos = TRUE,
                   selected.genes = commonMobGenes,
                   nTopGenes = NA,
                   genes.to.remove = NA,
                   perc.spots = NA,
                   min.reads = 100,
                   min.lib.size = 100,
                   ODgenes = FALSE,
                   od.genes.alpha = 0.05, # ODgenes ignored; so is this
                   gam.k = 5) # ODgenes ignored; so is this
  dat
})

names(mobCorpus) <- unlist(strsplit(stahlMobDatPaths, "_MOB_count_matrix-1.tsv"))

```

```{r}

# now do the MERINGUE mob. Positions added in separately
mobCorpus[["mob"]] <- preprocess(mob_paths[[1]],
                                 alignFile = NA,
                                 extractPos = FALSE,
                                 selected.genes = commonMobGenes,
                                 nTopGenes = NA,
                                 genes.to.remove = NA,
                                 perc.spots = NA,
                                 min.reads = 100,
                                 min.lib.size = 100,
                                 ODgenes = FALSE,
                                 od.genes.alpha = 0.05, # ODgenes ignored; so is this
                                 gam.k = 5) # ODgenes ignored; so is this

# positions already included in `mOB$pos` but make sure only use the filtered positions in the corpus
mobCorpus[["mob"]]$pos <- mOB$pos[rownames(mobCorpus[["mob"]]$corpus), ]

```

## MOB scRNAseq ref

```{r}

load("/Users/brendan/Desktop/PostDoc/work/STDeconvolve/data/GSE121891_mOB_scRNAseq/mob_processed_sparseMatrix.RData")
# mob_se

mob_meta <- read.csv2("/Users/brendan/Desktop/PostDoc/work/STDeconvolve/data/GSE121891_mOB_scRNAseq/GSE121891_OB_metaData_seurat.csv", sep = ",")

wt_cells <- mob_meta[which(mob_meta$orig.ident %in% c("WT1", "WT2")),]$"X"
print(length(wt_cells))

mob_se_wt <- mob_se[,wt_cells]

meta.data.wt = mob_meta[which(mob_meta$orig.ident %in% c("WT1", "WT2")),]

```

Construct proxy "beta" by combining gene expression of cells of same cell types

```{r}

mobWtCellClusters <- meta.data.wt$ClusterName
names(mobWtCellClusters) <- meta.data.wt$X

mobWtCellProxyBeta <- model.matrix(~ 0 + as.factor(mobWtCellClusters))

rownames(mobWtCellProxyBeta) <- names(mobWtCellClusters)

# fix names
colnames(mobWtCellProxyBeta) <- unlist(lapply(colnames(mobWtCellProxyBeta), function(x) {
  unlist(strsplit(x, ")"))[2]
}))

mobWtCellProxyBeta <- t(as.matrix(mob_se_wt) %*% mobWtCellProxyBeta)

mobWtCellProxyBeta <- mobWtCellProxyBeta/rowSums(mobWtCellProxyBeta)

dim(mobWtCellProxyBeta)

```



# Figure 1

```{r}

fig_path <- "/Users/brendan/Desktop/PostDoc/work/STDeconvolve/repos/STdeconvolve/figures/"

```


# A - Merfish mPOA (-0.04)

Show the individual cells in the bregma colored by the major cell class and overlay the simulated spot boundaries

```{r}

breg <- simBregmas$`-0.04`

ggplot() +
  geom_point(data = breg$annotDf,
             aes(x = Centroid_X, y = Centroid_Y, color = Cell_class)) +
  scale_fill_manual(values = breg$classColors) +
  theme(
        #panel.background = element_rect(fill = "white"),
        panel.grid = element_blank(),
        axis.line=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks=element_blank(),
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        panel.background=element_blank())
ggsave(filename = filename = "Fig1_A-1_breg04_mjrcts.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

ggplot() +
  geom_point(data = breg$annotDf,
             aes(x = Centroid_X, y = Centroid_Y, color = Cell_class)) +
  scale_fill_manual(values = breg$classColors) +
  geom_rect(data = breg$cellCounts,
            aes(xmin = x - 0, xmax = x + 100,
                ymin = y - 0, ymax = y + 100),
            fill = NA, color = "black") +
  theme(
        #panel.background = element_rect(fill = "white"),
        panel.grid = element_blank(),
        axis.line=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks=element_blank(),
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        panel.background=element_blank())
ggsave(filename = "Fig1_A-2_breg04_mjrcts_patches.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

# for all neuronal cell types 
# 72 types now so legend and assigned colors gets messy
breg <- simBregmas_neuro$`-0.04`

ggplot() +
  geom_point(data = breg$annotDf,
             aes(x = Centroid_X, y = Centroid_Y, color = Cell_class)) +
  scale_fill_manual(values = breg$classColors) +
  theme(
        #panel.background = element_rect(fill = "white"),
        panel.grid = element_blank(),
        axis.line=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks=element_blank(),
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        panel.background=element_blank(),
        legend.position = "none")
ggsave(filename = "Fig1_A-3_breg04_neurocts.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

ggplot() +
  geom_point(data = breg$annotDf,
             aes(x = Centroid_X, y = Centroid_Y, color = Cell_class)) +
  scale_fill_manual(values = breg$classColors) +
  geom_rect(data = breg$cellCounts,
            aes(xmin = x - 0, xmax = x + 100,
                ymin = y - 0, ymax = y + 100),
            fill = NA, color = "black") +
  theme(
        #panel.background = element_rect(fill = "white"),
        panel.grid = element_blank(),
        axis.line=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks=element_blank(),
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        panel.background=element_blank(),
        legend.position = "none")
ggsave(filename = "Fig1_A-4_breg04_neurocts_patches.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

```

# B - LDA results

Fit LDA models

```{r}

# fit models to a range of K's that include 9 (for 9 major cell types) and
# 75 (for all the cell types including neuronal)
ks <- seq(from = 2, to = 10, by = 1)
ks <- c(ks, seq(from = 15, to = 75, by = 10))

```

Recall that `simBregmas$`-0.04`$sim` and `simBregmas_neuro$`-0.04`$sim` have the same corpus, but the assigned ground truths constructed in "buildBregmaCorpus" differ because one used the 9 major cts and the other used major plus extended neuros.

So can use either "simBregmas" or "simBregmas_neuro" input corpuses to fit.

# -0.04

```{r}

start_time <- Sys.time()

pdf(file = paste0(fig_path, "Fig1_B-1_breg04_mjrcts_lda_fit.pdf"))
bregma04_LDAs <- fitLDA(counts = simBregmas$`-0.04`$sim,
                        Ks = ks,
                        seed = 0,
                        ncores = 7,
                        plot = TRUE)
dev.off()

total_t <- round(difftime(Sys.time(), start_time, units = "mins"), 2)
print(sprintf("Time to train LDA models was %smins", total_t))

```

Get beta, theta matrices and clusters of topics for model assessment and visualization

## k = 9 major cts

```{r}

bregma04_k9 <- buildLDAobject(LDAmodel = optimalModel(bregma04_LDAs, opt = 9),
                      deepSplit = 4,
                      colorScheme = "rainbow")

```

## viz predictions

```{r}

set.seed(888) # for randomColorR

# consistent colors between plots for topics
colspace <- randomcoloR::distinctColorPalette(9)

```

```{r}

m <- bregma04_k9$theta # theta for the k=9 fitted LDA for bregma -0.04
p <- simBregmas$`-0.04`$cellCounts[,c("x", "y")] # the positions of sim spots for the bregma -0.04

vizAllTopics(theta = m,
             pos = p,
             topicOrder=seq(ncol(m)),
             topicCols=colspace,
             groups = NA,
             group_cols = NA,
             r = 40,
             lwd = 0.5,
             showLegend = TRUE,
             plotTitle = NA)
ggsave(filename = "Fig1_B-2_breg04_k9_predict.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)


set.seed(1)

m <- simBregmas$`-0.04`$gtSpotTopics # The ground truth proportions for 9 major cts
p <- simBregmas$`-0.04`$cellCounts[,c("x", "y")] # the positions of sim spots for the bregma -0.04

vizAllTopics(theta = m,
             pos = p,
             topicOrder=seq(ncol(m)),
             topicCols=randomcoloR::distinctColorPalette(ncol(m)),
             groups = NA,
             group_cols = NA,
             r = 40,
             lwd = 0.5,
             showLegend = TRUE,
             plotTitle = NA)
ggsave(filename = "Fig1_B-3_breg04_gt_mjrcts.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

```

Individual topics visualized separately

```{r}

m <- bregma04_k9$theta # theta for the k=9 fitted LDA for bregma -0.04
p <- simBregmas$`-0.04`$cellCounts[,c("x", "y")] # the positions of sim spots for the bregma -0.04

cc <- as.factor(colspace)
names(cc) <- colnames(m)

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc,
                 sharedCol = TRUE,
                 groups = NA,
                 group_cols = NA,
                 r = 40,
                 lwd = 0.5,
                 showLegend = TRUE,
                 plotTitle = NA)

```

## theta correlation

bregma -0.04 fitted with k=9 vs gt 9 major cts

```{r}

thetaCor_breg04_k9_vs_gt <- getCorrMtx(m1 = bregma04_k9$theta,
                                       m2 = as.matrix(simBregmas$`-0.04`$gtSpotTopics),
                                       type = "t")

pdf(file = paste0(fig_path, "Fig1_B-4_breg04_k9_mjrcts_thetaCor.pdf"))
gplots::heatmap.2(x = thetaCor_breg04_k9_vs_gt,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Major cell class",
                  ylab = "Predicted topics",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(7,4))
dev.off()

pdf(file = paste0(fig_path, "Fig1_B-4_breg04_k9_mjrcts_thetaCor_pairs.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(thetaCor_breg04_k9_vs_gt)
gplots::heatmap.2(x = thetaCor_breg04_k9_vs_gt[pairs$rowix, pairs$colsix],
                  Rowv = NA,
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Major cell class",
                  ylab = "Predicted topics",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(7,4))
dev.off()

```

## beta correlation

bregma -0.04 fitted with k=9 vs gt 9 major cts

```{r}

betaCor_breg04_k9_vs_gt <- getCorrMtx(m1 = bregma04_k9$beta,
                                       m2 = as.matrix(simBregmas$`-0.04`$gtCtGenes), # gt made from cells in bregma
                                       type = "b")

pdf(file = paste0(fig_path, "Fig1_B-5_breg04_k9_mjrcts_betaCor.pdf"))
gplots::heatmap.2(x = betaCor_breg04_k9_vs_gt,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Major cell class",
                  ylab = "Predicted topics",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(7,4))
dev.off()

pdf(file = paste0(fig_path, "Fig1_B-5_breg04_k9_mjrcts_betaCor_pairs.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(betaCor_breg04_k9_vs_gt)
gplots::heatmap.2(x = betaCor_breg04_k9_vs_gt[pairs$rowix, pairs$colsix],
                  Rowv = NA,
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Major cell class",
                  ylab = "Predicted topics",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(7,4))
dev.off()

```

Compare with global gt for all of the cells

```{r}

betaCor_breg04_k9_vs_globalgt <- getCorrMtx(m1 = bregma04_k9$beta,
                                       m2 = as.matrix(gtCtGenes), # the global beta from all bregma cells in `bregmaCellsGexp` (36329)
                                       type = "b")

pdf(file = paste0(fig_path, "Fig1_B-5_breg04_k9_mjrcts_globalGt_betaCor.pdf"))
gplots::heatmap.2(x = betaCor_breg04_k9_vs_globalgt,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Major cell class",
                  ylab = "Predicted topics",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(7,4))
dev.off()

pdf(file = paste0(fig_path, "Fig1_B-5_breg04_k9_mjrcts_globalGt_betaCor_pairs.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(betaCor_breg04_k9_vs_globalgt)
gplots::heatmap.2(x = betaCor_breg04_k9_vs_globalgt[pairs$rowix, pairs$colsix],
                  Rowv = NA,
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Major cell class",
                  ylab = "Predicted topics",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(7,4))
dev.off()

```

## k = 75 all neuronal cts

```{r}

bregma04_k75 <- buildLDAobject(LDAmodel = optimalModel(bregma04_LDAs, opt = "min"),
                      deepSplit = 2, # collapsed more to reduce number of clusters to 17
                      colorScheme = "rainbow")

```

## viz predictions

```{r}

set.seed(888) # for randomColorR

# consistent colors between plots for topics
colspace <- randomcoloR::distinctColorPalette(75)

```

### individual topics

```{r}

m <- bregma04_k75$theta # theta for the k=75 fitted LDA for bregma -0.04
p <- simBregmas_neuro$`-0.04`$cellCounts[,c("x", "y")] # the positions of sim spots for the bregma -0.04

# Note: all topics are plotted regardless of proportion. So even topics at very residuals small amounts are plotted
# in every spot. It might be worth removing topics from specific spots if they are below a certain proportion

vizAllTopics(theta = m,
             pos = p,
             topicOrder=seq(ncol(m)),
             topicCols=colspace,
             groups = NA,
             group_cols = NA,
             r = 40,
             lwd = 0.01, # many more topics so make lines very thin
             showLegend = FALSE,
             plotTitle = "bregma -0.04 k=75 topics")
ggsave(filename = "Fig1_B-2_breg04_k75_predict.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)


set.seed(1)

m <- simBregmas_neuro$`-0.04`$gtSpotTopics # The ground truth proportions for the actual 72 of 75 present cts in bregma
p <- simBregmas_neuro$`-0.04`$cellCounts[,c("x", "y")] # the positions of sim spots for the bregma -0.04

vizAllTopics(theta = m,
             pos = p,
             topicOrder=seq(ncol(m)),
             topicCols=randomcoloR::distinctColorPalette(ncol(m)),
             groups = NA,
             group_cols = NA,
             r = 40,
             lwd = 0.01,
             showLegend = FALSE,
             plotTitle = "Ground truth proportions 75 cts")
ggsave(filename = "Fig1_B-3_breg04_gt_75neurocts.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

```

Because so many topics, collapse the individual topics into topic clusters to better visualize and assess

```{r}

set.seed(888) # for randomColorR

# consistent colors between plots for topics
colspace <- randomcoloR::distinctColorPalette(ncol(bregma04_k75$thetaCombn))

```

### topic clusters

```{r}

m <- bregma04_k75$thetaCombn # theta of the combined topic clusters for the k=75 fitted LDA for bregma -0.04
p <- simBregmas_neuro$`-0.04`$cellCounts[,c("x", "y")] # the positions of sim spots for the bregma -0.04

# Note: all topics are plotted regardless of proportion. So even topics at very residuals small amounts are plotted
# in every spot. It might be worth removing topics from specific spots if they are below a certain proportion

vizAllTopics(theta = m,
             pos = p,
             topicOrder=seq(ncol(m)),
             topicCols=colspace,
             groups = NA,
             group_cols = NA,
             r = 40,
             lwd = 0.01, # many more topics so make lines very thin
             showLegend = FALSE,
             plotTitle = "bregma -0.04 k=75 topic-clusters")
ggsave(filename = "Fig1_B-2_breg04_k75_clusters_predict.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

```

Lets visualize each topic cluster separately but also observe the proportions of the individual topics that make up each cluster. For example, a topic cluster clearly captures the ependymal cells, but there is no single individual topic that captures this. Instead, the LDA appears to capture additional transcriptionally distinct subtypes of the ependymal cells.

```{r}

m <- bregma04_k75$theta # theta for the k=75 fitted LDA for bregma -0.04
p <- simBregmas_neuro$`-0.04`$cellCounts[,c("x", "y")] # the positions of sim spots for the bregma -0.04

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = bregma04_k75$cols, # factor for the 75 topics where the levels are the color of assigned cluster
                 sharedCol = FALSE, # so that each topic in a cluster is colored as a shade of the cluster color
                 groups = NA,
                 group_cols = NA,
                 r = 40,
                 lwd = 0.5,
                 showLegend = TRUE,
                 plotTitle = NA)

```

Break down even more but just looking at specific individual clusters

```{r}

m <- bregma04_k75$theta[,c(22,23,32,41,46,60)] # theta for the k=9 fitted LDA for bregma -0.04
p <- simBregmas_neuro$`-0.04`$cellCounts[,c("x", "y")] # the positions of sim spots for the bregma -0.04

cc <- as.factor(rainbow(ncol(m)))
names(cc) <- colnames(m)

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc,
                 sharedCol = TRUE,
                 groups = NA,
                 group_cols = NA,
                 r = 40,
                 lwd = 0.5,
                 showLegend = TRUE,
                 plotTitle = NA)

```

# C - Example ependymal

# bregma -0.04 k=9 mjr cts

```{r}

# Topic 7 highly correlated with ependymal cells
topic7ependymalTopGenes <- head(bregma04_k9$beta[7,order(bregma04_k9$beta[7,], decreasing = TRUE)], n=10)
topic7ependymalTopGenes

```

Visualize genes:

```{r}

# input df for `vizGeneCounts`
df <- bregmaCellsGexp[,labels(topic7ependymalTopGenes)]
df <- merge(as.data.frame(spatial_position_and_class[,c("Centroid_X", "Centroid_Y")]), as.data.frame(df), by=0)
colnames(df)[which(names(df) == "Centroid_X")] <- "x"
colnames(df)[which(names(df) == "Centroid_Y")] <- "y"
rownames(df) <- df$Row.names

```

```{r}

vizGeneCounts(df = df,
              gene = "Mlc1",
              groups = NA,
              group_cols = NA,
              size = 0.5, stroke = 0.01,
              plotTitle = "Mlc1",
              showLegend = TRUE)
ggsave(filename = "Fig1_C-1_breg04_k9_ependy_gexp_Mlc1.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

vizGeneCounts(df = df,
              gene = "Cd24a",
              groups = NA,
              group_cols = NA,
              size = 0.5, stroke = 0.01,
              plotTitle = "Cd24a",
              showLegend = TRUE)

vizGeneCounts(df = df,
              gene = "Aqp4",
              groups = NA,
              group_cols = NA,
              size = 0.5, stroke = 0.01,
              plotTitle = "Aqp4",
              showLegend = TRUE)

```

# bregma -0.04 k=75 neuro cts

Above, it was observed that the ependymal cells are captured by the topic cluster of the following topics:
F0FF00 : 22 23 32 41 46 60

When the topics are visualized separately, the individual topics appear to capture additional transcriptionally distinct subpopulations of the ependymal cells.

Let's visualize the top genes for the cluster in general, and also for the individual topics

```{r}

bregma04_k75$clusters

```

Note that topics "22 23 32 41 46 60" are part of cluster `4`

```{r}

breg04_k75_ependymalTopGenes_tclust <- head(bregma04_k75$betaCombn[4,order(bregma04_k75$betaCombn[4,], decreasing = TRUE)], n=10)
breg04_k75_ependymalTopGenes_tclust

```

The genes for the overall cluster are very similar ro what was found with k=9. So representative of ependymal cells overall.

But what about the individual topics in the cluster?

```{r}

print("Topic 22")
head(bregma04_k75$beta[22,order(bregma04_k75$beta[22,], decreasing = TRUE)], n=10)

print("Topic 23")
head(bregma04_k75$beta[23,order(bregma04_k75$beta[23,], decreasing = TRUE)], n=10)

print("Topic 32")
head(bregma04_k75$beta[32,order(bregma04_k75$beta[32,], decreasing = TRUE)], n=10)

print("Topic 41")
head(bregma04_k75$beta[41,order(bregma04_k75$beta[41,], decreasing = TRUE)], n=10)

print("Topic 46")
head(bregma04_k75$beta[46,order(bregma04_k75$beta[46,], decreasing = TRUE)], n=10)

print("Topic 60")
head(bregma04_k75$beta[60,order(bregma04_k75$beta[60,], decreasing = TRUE)], n=10)


```

Again, Mlc1 is a top gene for all of them. Cd24a also top marker.  

```{r}

genes <- c("Mlc1", "Cd24a", "Aqp4", "Cd24a", "Cxcl14", "Sema4d", "Ndrg1")

# input df for `vizGeneCounts`
df <- bregmaCellsGexp[,genes]
df <- merge(as.data.frame(spatial_position_and_class[,c("Centroid_X", "Centroid_Y")]), as.data.frame(df), by=0)
colnames(df)[which(names(df) == "Centroid_X")] <- "x"
colnames(df)[which(names(df) == "Centroid_Y")] <- "y"
rownames(df) <- df$Row.names

```

Topic 32 appears to be very specific at the bottom of the ependymal layer. It is the strongest expression wrt Mlc1 compared to the other topics and also has "Cd24a", "Sema4d" and "Ndrg1"

It's possible that this topic is picking up more of an "artifact" for bregma -0.04 specifically as Mcl1 is particularly high in this region compared to other bregma slices. However, the possibility remains that this is a distinct subpopulation for this particular slice in space.

```{r}

vizGeneCounts(df = df,
              gene = "Mlc1",
              groups = NA,
              group_cols = NA,
              size = 0.5, stroke = 0.01,
              plotTitle = "Mlc1",
              showLegend = TRUE)
# ggsave(filename = "Fig1_C-2_breg04_k75_ependy_gexp_Slco1a4.pdf",
#        device = "pdf",
#        path = fig_path,
#        scale = 1.5,
#        width = 5,
#        height = 4,
#        units = c("in"),
#        dpi = 600)

vizGeneCounts(df = df,
              gene = "Cd24a",
              groups = NA,
              group_cols = NA,
              size = 0.5, stroke = 0.01,
              plotTitle = "Cd24a",
              showLegend = TRUE)

vizGeneCounts(df = df,
              gene = "Sema4d",
              groups = NA,
              group_cols = NA,
              size = 0.5, stroke = 0.01,
              plotTitle = "Sema4d",
              showLegend = TRUE)

vizGeneCounts(df = df,
              gene = "Ndrg1",
              groups = NA,
              group_cols = NA,
              size = 0.5, stroke = 0.01,
              plotTitle = "Ndrg1",
              showLegend = TRUE)

vizGeneCounts(df = df,
              gene = "Cxcl14",
              groups = NA,
              group_cols = NA,
              size = 0.5, stroke = 0.01,
              plotTitle = "Cxcl14",
              showLegend = TRUE)

```

# D - MOB 

Original Txn clusters

```{r, mob-qc, fig.width=8, fig.height=3}

# Remove poor datasets and genes
mob_counts <- MERINGUE::cleanCounts(counts = mOB$counts, # from mOB data set
                      min.reads = 100, 
                      min.lib.size = 100, 
                      plot=TRUE,
                      verbose=TRUE)
mob_pos <- mOB$pos[colnames(mob_counts),]

# CPM normalize
mob_cpm <- MERINGUE::normalizeCounts(counts = mob_counts, 
                       log=FALSE,
                       verbose=TRUE)

```

```{r}

# Dimensionality reduction by PCA on log10 CPM expression values
pcs.info <- prcomp(t(log10(as.matrix(mob_cpm)+1)), center=TRUE)
nPcs <- 5
pcs <- pcs.info$x[,1:nPcs]

# 2D embedding by tSNE
emb <- Rtsne::Rtsne(pcs,
             is_distance=FALSE,
             perplexity=30,
             num_threads=1,
             verbose=FALSE)$Y
rownames(emb) <- rownames(pcs)

# Graph-based cluster detection
k <- 30
com <- getClusters(pcs, k, weight=TRUE)

# Manually annotate identified clusters with cell-types
mob_annot <- as.character(com); names(mob_annot) <- names(com)
mob_annot[com==4] <- '1: Granule Cell Layer'
mob_annot[com==1] <- '2: Mitral Cell Layer'
mob_annot[com==3] <- '3: Outer Plexiform Layer'
mob_annot[com==2] <- '4: Glomerular Layer'
mob_annot[com==5] <- '5: Olfactory Nerve Layer'
mob_annot <- as.factor(mob_annot)

# Plot
par(mfrow=c(2,2), mar=rep(1,4))
plotEmbedding(emb, groups=mob_annot, 
              show.legend=TRUE, xlab=NA, ylab=NA,
              verbose=FALSE)
plotEmbedding(mob_pos, groups=mob_annot, 
              cex=1, xlab=NA, ylab=NA,
              verbose=FALSE)

```

# E - LDA results

Fit LDA models

```{r}

# fit models to a range of K's that include 5 (5 original txn clusters) and
# 37 and 38 (for the cts in the scRNAseq refs; 38 originally but SL doesn't find and Meis2 so dropped)
ks <- seq(from = 2, to = 10, by = 1)
ks <- c(ks, 37, 38, seq(from = 15, to = 75, by = 10))

```

Recall that `simBregmas$`-0.04`$sim` and `simBregmas_neuro$`-0.04`$sim` have the same corpus, but the assigned ground truths constructed in "buildBregmaCorpus" differ because one used the 9 major cts and the other used major plus extended neuros.

So can use either "simBregmas" or "simBregmas_neuro" input corpuses to fit.

# mob (meringue data)

```{r}

pdf(file = paste0(fig_path, "Fig1_D-1_mob_lda_fit.pdf"))
mob_LDAs <- fitLDA(counts = mobCorpus$mob$slm,
                        Ks = ks,
                        seed = 0,
                        ncores = 7,
                        plot = TRUE)
dev.off()

```

## k = 15 kOpt2 kneed

```{r}

mob_k15 <- buildLDAobject(LDAmodel = optimalModel(mob_LDAs, opt = "kneed"),
                      deepSplit = 4,
                      colorScheme = "rainbow")

```

## viz predictions

### individual topics

```{r}

set.seed(888) # for randomColorR

# consistent colors between plots for topics
colspace <- randomcoloR::distinctColorPalette(15)

```

```{r}

m <- mob_k15$theta # theta for the k=15 fitted LDA for mob data set
p <- mobCorpus$mob$pos # the positions of filtered spots for mob data set

vizAllTopics(theta = m,
             pos = p,
             topicOrder=seq(ncol(m)),
             topicCols=rainbow(ncol(m)),
             groups = NA,
             group_cols = NA,
             r = 0.4, # different size piecharts for mOB data sets
             lwd = 0.01,
             showLegend = TRUE,
             plotTitle = NA)
ggsave(filename = "Fig1_E-1_mob_k15_predict.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

```

### topic clusters

```{r}

set.seed(888) # for randomColorR

# consistent colors between plots for topics
colspace <- randomcoloR::distinctColorPalette(6)

```

```{r}

m <- mob_k15$thetaCombn # theta for 6 topic clusters of the k=15 fitted LDA for mob data set
p <- mobCorpus$mob$pos # the positions of filtered spots for mob data set

vizAllTopics(theta = m,
             pos = p,
             topicOrder=seq(ncol(m)),
             topicCols=rainbow(ncol(m)),
             groups = NA,
             group_cols = NA,
             r = 0.4, # different size piecharts for mOB data sets
             lwd = 0.01,
             showLegend = TRUE,
             plotTitle = NA)
ggsave(filename = "Fig1_E-2_mob_k15_clusters_predict.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

```

### topics separately

Visualize the individual topics and also the topic clusters separately

```{r}

m <- mob_k15$theta
p <- mobCorpus$mob$pos

cc <- as.factor(rainbow(ncol(m)))
names(cc) <- colnames(m)

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc,
                 sharedCol = TRUE,
                 groups = NA,
                 group_cols = NA,
                 r = 0.4,
                 lwd = 0.01,
                 showLegend = TRUE,
                 plotTitle = NA)

```

```{r}

m <- mob_k15$thetaCombn
p <- mobCorpus$mob$pos

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = mob_k15$clustCols,
                 sharedCol = TRUE,
                 groups = NA,
                 group_cols = NA,
                 r = 0.4,
                 lwd = 0.01,
                 showLegend = TRUE,
                 plotTitle = NA)

```

Check out the individual topics that make up each cluster

```{r}

m <- mob_k15$theta
p <- mobCorpus$mob$pos

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = mob_k15$cols,
                 sharedCol = FALSE,
                 groups = NA,
                 group_cols = NA,
                 r = 0.4,
                 lwd = 0.01,
                 showLegend = TRUE,
                 plotTitle = NA)

```

## theta correlation

See which topics correlate with the 5 cell layer annotations based on txn clustering

```{r}

# proxy theta for the txn clusters
mobProxyTheta <- model.matrix(~ 0 + mob_annot)
rownames(mobProxyTheta) <- names(mob_annot)

# fix names
colnames(mobProxyTheta) <- unlist(lapply(colnames(mobProxyTheta), function(x) {
  unlist(strsplit(x, ": "))[2]
}))

mobProxyTheta <- as.data.frame.matrix(mobProxyTheta)

```

```{r}

thetaCor_mob_k15_vs_gt <- getCorrMtx(m1 = as.matrix(mobProxyTheta),
                                       m2 = mob_k15$theta,
                                       type = "t")

pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor.pdf"))
gplots::heatmap.2(x = thetaCor_mob_k15_vs_gt,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Predicted topics",
                  ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(4,11)) # margins for labels
dev.off()

pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor_pairs.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(thetaCor_mob_k15_vs_gt)
gplots::heatmap.2(x = thetaCor_mob_k15_vs_gt[pairs$rowix, pairs$colsix],
                  Rowv = NA,
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Predicted topics",
                  ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(4,11))
dev.off()

```

Topic cluster theta correlations

```{r}

thetaCor_combined_mob_k15_vs_gt <- getCorrMtx(m1 = as.matrix(mobProxyTheta),
                                       m2 = mob_k15$thetaCombn,
                                       type = "t")

pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_theta_combined_Cor.pdf"))
gplots::heatmap.2(x = thetaCor_combined_mob_k15_vs_gt,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Predicted topics",
                  ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(4,11)) # margins for labels
dev.off()

pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_theta_combined_Cor_pairs.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(thetaCor_combined_mob_k15_vs_gt)
gplots::heatmap.2(x = thetaCor_combined_mob_k15_vs_gt[pairs$rowix, pairs$colsix],
                  Rowv = NA,
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Predicted topics",
                  ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(4,11))
dev.off()

```

## beta correlation

Correlate to a mob scRNAseq reference to see if certain topics also correlate with known cell types.

```{r}

betaCor_mob_k15_vs_gt <- getCorrMtx(m1 = mob_k15$beta,
                                       m2 = as.matrix(mobWtCellProxyBeta), # gt proxy beta
                                       type = "b")

pdf(file = paste0(fig_path, "Fig1_E-4_mob_k15_scRNAseqProxy_betaCor.pdf"))
gplots::heatmap.2(x = betaCor_mob_k15_vs_gt,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Major cell class",
                  ylab = "Predicted topics",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,4))
dev.off()

pdf(file = paste0(fig_path, "Fig1_E-4_mob_k15_scRNAseqProxy_betaCor_pairs.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(betaCor_mob_k15_vs_gt)
gplots::heatmap.2(x = betaCor_mob_k15_vs_gt[pairs$rowix, pairs$colsix],
                  Rowv = NA,
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Predicted topics",
                  ylab = "Major cell class",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,4))
dev.off()

```

```{r}

betaCor_mob_k15_combined_vs_gt <- getCorrMtx(m1 = mob_k15$betaCombn,
                                       m2 = as.matrix(mobWtCellProxyBeta), # gt proxy beta
                                       type = "b")

pdf(file = paste0(fig_path, "Fig1_E-5_mob_k15_scRNAseqProxy_betaCor_combined.pdf"))
gplots::heatmap.2(x = betaCor_mob_k15_combined_vs_gt,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Major cell class",
                  ylab = "Predicted topics",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,4))
dev.off()

pdf(file = paste0(fig_path, "Fig1_E-5_mob_k15_scRNAseqProxy_betaCor_combined_pairs.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(betaCor_mob_k15_combined_vs_gt)
gplots::heatmap.2(x = betaCor_mob_k15_combined_vs_gt[pairs$rowix, pairs$colsix],
                  Rowv = NA,
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "Predicted topics",
                  ylab = "Major cell class",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,4))
dev.off()

```

# F - cell layer example

Can start to annotate topics based on where they are spatially and how they correlate with known cell types. Viz marker genes and also orthogonal validation with ISH data.



# Figure 2

lda k equal to other methods

merfish assess all cell types - which cell types do methods struggle most with? rare ones?

have merfish "scRNAseq" ref that you can use to assess what happens when a ct is removed for SL and RCTD

consider mob scRNAseq ref as gt and compute correlation. best matched and do RMSE?


beta probabilty versus actual gene counts...scatter plot of predicted beta vs actual cell type gene counts

there is that Interactive correlation heatmap might be useful here.

For example, the W vs the scRNAseq gene exp correlation. I did this at some point. Revisit the correlations by seeing the individual genes. Does transforming gene counts make them correlate better with beta/W?







