---
title: "Untitled"
author: "Brendan F. Miller"
date: "3/23/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}

# all of the trained models for mob:
save(mob_k15, # LDA model of the mob data with k=15 and 93 common OD genes across the 12 MOB samples
     mob_k38, # LDA model ofthe mob data with k=38 and 93 common OD genes across the 12 MOB samples
     mobCorpus, # corpus of all the mob datasets and also has the spot positions
     mobProxyTheta, # table of mob 260 spots and which of the 5 txn clusters it was assigned
     countsClean, # 260 x 7365 genes of mob corpus after removing bad spots and genes
     nmf_mod_mob_wt_countsClean_ls, # SL fitted model trained with log-norm `mob_se_wt_obj` and `countsClean` ST data
     sl_countsClean, # above model after cleanup. 6338 genes kept that were in ST data and the scRNAseq. 38 cts but Mural2 is 0
     RCTD_mob_results_norm, # RCTD model trained with `mob_se_wt_raw_obj` and deconvolved `countsClean` ST. 28 cts
     nmf_mod_mob_wt_countsClean_noOECs_ls, # SL trained with log-norm `mob_se_wt_obj` and `countsClean` ST data but no OEC so 6317 genes and 33 cts
     sl_countsClean_noOEC, # above model after cleanup.
     RCTD_mob_noOEC_results_norm, # RCTD # RCTD model trained with `mob_se_wt_raw_obj` and deconvolved `countsClean` ST. but 33 cts now
     sl_cortexRef_countsClean_ls, # SL on cortex SCT data and on `countsClean`. 7072 shared genes with ST and 23 cts
     sl_countsClean_cortexRef, # above model after cleanup. but L5.IT is 0
     RCTD_cortexRef_results_norm, # RCTD on raw counts of cortex scRNAseq and `countsClean`. 23 cts
     file = "/Users/brendan/Desktop/PostDoc/work/STDeconvolve/repos/STdeconvolve/data/mob_fitted_models.RData")

load("/Users/brendan/Desktop/PostDoc/work/STDeconvolve/repos/STdeconvolve/0draft_data/mob_fitted_models.RData")


# sl_ctTheta <- sl_countsClean$thetaCt[,which(!colnames(sl_countsClean$thetaCt) %in% c("Mural2"))]
# sl_countsClean$thetaCt but no Mural2 because 0 in all spots

# for assessing beta values
save(mob_se_wt, # processed data matrix for 17709 wt cells and 18560 genes (filtered genes during the processing)
     mob_se_wt_obj, # seurat object of the processed data and meta data
     mob_se_wt_raw, # raw counts of the 17709 wt cells and 18560 genes
     mob_se_wt_raw_obj_sct, # seurat object of raw data, and also my own SCT processing (but turned out different than the supplied data)
     cluster_markers_mob_se_wt, # Seurat::FindAllMarkers for the processed wt cells and 38 clusters
     cluster_markers_mob_se_wt_raw, # Seurat::FindAllMarkers for the raw counts wt cells and 38 clusters
     mobWtCellProxyBeta, # averaged relative counts for the 38 clusters for 18560 genes by averaging the processed log-norm data
     mobWtCellProxyBetaRaw, # averaged relative counts for the 38 clusters for 18560 genes by averaging the raw counts
     mob_k15, # LDA model of the mob data with k=15 and 93 common OD genes across the 12 MOB samples
     sl_countsClean, # SL model trained on `mob_se_wt_obj` and `countsClean` ST data of the `mOB` data (260 spots and 6336 shared genes)
     sl_ctTheta, # thetaCt of the above model, but Mural2 removed because 0 in all spots
     W_countsClean, # betaTopics of the sl_countsClean; the W matrix of the NMF fitted model
     file = "/Users/brendan/Desktop/PostDoc/work/STDeconvolve/repos/STdeconvolve/data/mob_data_to_assess_beta.RData")

load("/Users/brendan/Desktop/PostDoc/work/STDeconvolve/repos/STdeconvolve/0draft_data/mob_data_to_assess_beta.RData")

cortexProxyBetaRaw <- readRDS("/Users/brendan/Desktop/PostDoc/work/STDeconvolve/repos/STdeconvolve/0draft_data/cortexProxyBetaRaw.Rds")
# 23 29093 average raw gene counts from the cortex_sc seurat object. The 23 subclasses

```

```{r}

fig_path <- "/Users/brendan/Desktop/PostDoc/work/STDeconvolve/repos/STdeconvolve/figures/"
draft_fig_path <- "/Users/brendan/Desktop/PostDoc/work/STDeconvolve/repos/STdeconvolve/0draft_figures/"

```

# viz k=15

```{r}

m <- mob_k15$theta # theta for the k=15 fitted LDA for mob data set
p <- mobCorpus$mob$pos # the positions of filtered spots for mob data set

vizAllTopics(theta = m,
             pos = p,
             topicOrder=seq(ncol(m)),
             topicCols=gg_color_hue(ncol(m)),
             groups = rep("0", dim(m)[1]),
             group_cols = c("0" = "black"),
             r = 0.4, # different size piecharts for mOB data sets
             lwd = 0.1,
             showLegend = TRUE,
             plotTitle = NA)
ggsave(filename = "Fig_2_mob-k15.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

```

# Compare models

## LDA vs SL

```{r}

corMtx <- getCorrMtx(m1 = sl_ctTheta,
                         m2 = mob_k38$theta,
                         type = "t")

# pdf(file = paste0(fig_path, "Fig1_B-4_FN7_k8_mjrcts_thetaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Major cell class",
                  # ylab = "Predicted topics",
                  key.title = NA,
                  cexRow = 0.6,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(2,4))
# dev.off()

# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(corMtx)
# new mtx where rows and cols in order in which they are paired
corMtx_paired <- corMtx[pairs$rowix, pairs$colsix]

# get dendro clusters for the mob cts based on the gexp similarity
mob_sig_marker_genes <- unique(cluster_markers_mob_se_wt_raw[cluster_markers_mob_se_wt_raw$p_val_adj < 0.05,]$gene)
beta <- mobWtCellProxyBetaRaw[rownames(corMtx_paired), mob_sig_marker_genes]
# beta <- sl_countsClean$betaCt[colnames(corMtx_paired),]
d_ <- as.dist(1-cor(t(beta)))
hc_ <- stats::hclust(d_, method = "ward.D")

# pdf(file = paste0(fig_path, "Fig_S6_SL-mob-vs-cortex-corr-heatmap.pdf"))
gplots::heatmap.2(x = corMtx_paired[hc_$order,],
                  Rowv = NA, # to order topics based on pairs and not dendrogram
                  Colv = as.dendrogram(hc_),
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "STdevonvolve",
                  ylab = "SPOTlight",
                  key.title = NA,
                  cexRow = 0.6,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(4,5))
# dev.off()

lda_vs_sl_theta <- diag(corMtx_paired)

hist(lda_vs_sl_theta, breaks = 10)
mean(lda_vs_sl_theta)
sd(lda_vs_sl_theta)

median(lda_vs_sl_theta)
range(lda_vs_sl_theta)

# pdf(file = paste0(fig_path, "Fig_S6_SL-mob-vs-cortex-corr-boxplot.pdf"),   # The directory you want to save the file in
    # width = 5, # The width of the plot in inches
    # height = 3) # The height of the plot in inches
par( mfrow = c(1,3) )
boxplot(lda_vs_sl_theta,
        medcol = "black", # medium line color
        medlty = 1, # line type
        medlwd = 1, # line width
        boxfill = "white", # color fill of box
        boxcol = "black", # outline color of box
        boxlty = 1, # box line type
        boxlwd = 1, # width of box line
        whisklty = 1, # whisker type
        whisklwd = 1,
        staplelwd = 0, # the ends of the whiskers
        outpch = 1, # type of outlies
        outcex = 0,
        main = "",
        cex.main = 1,
        xlab = "",
        ylab = "Correlation") # size of outliers
stripchart(lda_vs_sl_theta, vertical = TRUE,
    method = "jitter", add = TRUE, pch = 20, col = 'black')
# dev.off()

```

## RCTD vs SL

```{r}

corMtx <- getCorrMtx(m1 = as.matrix(RCTD_mob_results_norm),
                     m2 = sl_ctTheta,
                     type = "t")

# pdf(file = paste0(fig_path, "Fig1_B-4_FN7_k8_mjrcts_thetaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "SL",
                  ylab = "RCTD",
                  key.title = NA,
                  cexRow = 0.6,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,5))
# dev.off()

# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(corMtx)
# new mtx where rows and cols in order in which they are paired
corMtx_paired <- corMtx[pairs$rowix, pairs$colsix]

# get dendro clusters for the mob cts based on the gexp similarity
mob_sig_marker_genes <- unique(cluster_markers_mob_se_wt_raw[cluster_markers_mob_se_wt_raw$p_val_adj < 0.05,]$gene)
beta <- mobWtCellProxyBetaRaw[rownames(corMtx_paired), mob_sig_marker_genes]
# beta <- sl_countsClean$betaCt[colnames(corMtx_paired),]
d_ <- as.dist(1-cor(t(beta)))
hc_ <- stats::hclust(d_, method = "ward.D")

# pdf(file = paste0(fig_path, "Fig_S6_SL-mob-vs-cortex-corr-heatmap.pdf"))
gplots::heatmap.2(x = corMtx_paired[hc_$order,],
                  Rowv = NA, # to order topics based on pairs and not dendrogram
                  Colv = as.dendrogram(hc_),
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "SL",
                  ylab = "RCTD",
                  key.title = NA,
                  cexRow = 0.6,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,5))
# dev.off()

rctd_vs_sl_theta <- diag(corMtx_paired)

hist(rctd_vs_sl_theta, breaks = 10)
mean(rctd_vs_sl_theta)
sd(rctd_vs_sl_theta)

median(rctd_vs_sl_theta)
range(rctd_vs_sl_theta)

# pdf(file = paste0(fig_path, "Fig_S6_SL-mob-vs-cortex-corr-boxplot.pdf"),   # The directory you want to save the file in
    # width = 5, # The width of the plot in inches
    # height = 3) # The height of the plot in inches
par( mfrow = c(1,3) )
boxplot(rctd_vs_sl_theta,
        medcol = "black", # medium line color
        medlty = 1, # line type
        medlwd = 1, # line width
        boxfill = "white", # color fill of box
        boxcol = "black", # outline color of box
        boxlty = 1, # box line type
        boxlwd = 1, # width of box line
        whisklty = 1, # whisker type
        whisklwd = 1,
        staplelwd = 0, # the ends of the whiskers
        outpch = 1, # type of outlies
        outcex = 0,
        main = "",
        cex.main = 1,
        xlab = "",
        ylab = "Correlation") # size of outliers
stripchart(rctd_vs_sl_theta, vertical = TRUE,
    method = "jitter", add = TRUE, pch = 20, col = 'black')
# dev.off()

```

## RCTD vs LDA

```{r}

corMtx <- getCorrMtx(m1 = as.matrix(RCTD_mob_results_norm),
                     m2 = mob_k38$theta,
                     type = "t")

# pdf(file = paste0(fig_path, "Fig1_B-4_FN7_k8_mjrcts_thetaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "STdeconvolve",
                  ylab = "RCTD",
                  key.title = NA,
                  cexRow = 0.6,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,5))
# dev.off()

# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(corMtx)
# new mtx where rows and cols in order in which they are paired
corMtx_paired <- corMtx[pairs$rowix, pairs$colsix]

# get dendro clusters for the mob cts based on the gexp similarity
mob_sig_marker_genes <- unique(cluster_markers_mob_se_wt_raw[cluster_markers_mob_se_wt_raw$p_val_adj < 0.05,]$gene)
beta <- mobWtCellProxyBetaRaw[rownames(corMtx_paired), mob_sig_marker_genes]
# beta <- sl_countsClean$betaCt[colnames(corMtx_paired),]
d_ <- as.dist(1-cor(t(beta)))
hc_ <- stats::hclust(d_, method = "ward.D")

# pdf(file = paste0(fig_path, "Fig_S6_SL-mob-vs-cortex-corr-heatmap.pdf"))
gplots::heatmap.2(x = corMtx_paired[hc_$order,],
                  Rowv = NA, # to order topics based on pairs and not dendrogram
                  Colv = as.dendrogram(hc_),
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "STdeconvolve",
                  ylab = "RCTD",
                  key.title = NA,
                  cexRow = 0.6,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,5))
# dev.off()

rctd_vs_lda_theta <- diag(corMtx_paired)

hist(rctd_vs_lda_theta, breaks = 10)
mean(rctd_vs_lda_theta)
sd(rctd_vs_lda_theta)

median(rctd_vs_lda_theta)
range(rctd_vs_lda_theta)

# pdf(file = paste0(fig_path, "Fig_S6_SL-mob-vs-cortex-corr-boxplot.pdf"),   # The directory you want to save the file in
    # width = 5, # The width of the plot in inches
    # height = 3) # The height of the plot in inches
par( mfrow = c(1,3) )
boxplot(rctd_vs_lda_theta,
        medcol = "black", # medium line color
        medlty = 1, # line type
        medlwd = 1, # line width
        boxfill = "white", # color fill of box
        boxcol = "black", # outline color of box
        boxlty = 1, # box line type
        boxlwd = 1, # width of box line
        whisklty = 1, # whisker type
        whisklwd = 1,
        staplelwd = 0, # the ends of the whiskers
        outpch = 1, # type of outlies
        outcex = 0,
        main = "",
        cex.main = 1,
        xlab = "",
        ylab = "Correlation") # size of outliers
stripchart(rctd_vs_lda_theta, vertical = TRUE,
    method = "jitter", add = TRUE, pch = 20, col = 'black')
# dev.off()

```

## summarize

```{r}

cor_mob_summary <- list("STdeconvolve vs RCTD" = rctd_vs_lda_theta,
                        "RCTD vs SL" = rctd_vs_sl_theta,
                        "STdeconvolve vs SL" = lda_vs_sl_theta)

# reshape2::melt(plyr::ldply(cor_mob_summary, rbind))

# my_comparisons <- list( c("lda_vs_sl", "lda_vs_rctd"), c("lda_vs_rctd", "sl_vs_rctd"), c("lda_vs_sl", "sl_vs_rctd") )

# `plyr::ldply` combines vectors in list of different lenghts into dataframe
# then `reshape2::melt` into long form dataframe for ggplot

ggplot(data = reshape2::melt(plyr::ldply(cor_mob_summary, rbind)), 
       aes(x=.id, y=value, fill=.id)) +
  geom_boxplot() + 
  # stat_compare_means(comparisons = my_comparisons, method = "wilcox.test",
  #                    label.y = c(1.02, 1.09, 1.19)) +
  geom_point(position=position_jitterdodge()) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(title = "Method Correlations between best matched predicted cell types",
       x = "Method", y = "Correlation")
ggsave(filename = "Fig_2_mob-methods-corr-each-other.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

```

Maybe SL and RCTD correlate with each other because using same scRNAseq ref? But some cell types correlate poorly. Which ones? Are they predicted to be at low proportions overall?

```{r}

bottom <- order(rctd_vs_sl_theta, decreasing = FALSE)[1:5]

corMtx <- getCorrMtx(m1 = as.matrix(RCTD_mob_results_norm), # rows
                           m2 = sl_ctTheta, # columns
                           type = "t")
pairs <- lsatPairs(corMtx)

rctd_order <- rownames(corMtx[pairs$rowix, pairs$colsix])
rctd_means <- colMeans(as.matrix(RCTD_mob_results_norm))[rctd_order]

sl_order <- colnames(corMtx[pairs$rowix, pairs$colsix])
sl_means <- colMeans(sl_ctTheta)[sl_order]

df <- data.frame("RCTD mean" = rctd_means,
                 "SL mean" = sl_means,
                 "RCTD ct" = rctd_order,
                 "SL ct" = sl_order,
                 "correlation" = rctd_vs_sl_theta,
                 "rank" = rank(rctd_vs_sl_theta))

corMtx[pairs$rowix[bottom], pairs$colsix[bottom]]
df

ggplot(data = df) +
  geom_point(aes(x = RCTD.mean, y = SL.mean, color = correlation)) +
  scale_color_viridis(option = "A", direction = -1) +
  theme_classic() +
  # theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(title = "Correlation of predicted proportions of paired cell types",
       x = "RCTD mean spot proportion", y = "SPOTlight mean spot proportion")
ggsave(filename = "Fig_S_mob-RCTD-SL-ct_corrs.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

```

Recall that RCTD returned only 28 detected cts whereas SPOTlight returned predictions for all 38.

```{r}

bottom <- order(rctd_vs_lda_theta, decreasing = FALSE)[1:5]

corMtx <- getCorrMtx(m1 = as.matrix(RCTD_mob_results_norm), # rows
                           m2 = mob_k38$theta, # columns
                           type = "t")
pairs <- lsatPairs(corMtx)

rctd_order <- rownames(corMtx[pairs$rowix, pairs$colsix])
rctd_means <- colMeans(as.matrix(RCTD_mob_results_norm))[rctd_order]

lda_order <- colnames(corMtx[pairs$rowix, pairs$colsix])
lda_means <- colMeans(mob_k38$theta)[lda_order]

df <- data.frame("RCTD mean" = rctd_means,
                 "STdeconvolve mean" = lda_means,
                 "RCTD ct" = rctd_order,
                 "STdeconvolve ct" = lda_order,
                 "correlation" = rctd_vs_lda_theta,
                 "rank" = rank(rctd_vs_lda_theta))

corMtx[pairs$rowix[bottom], pairs$colsix[bottom]]
df

ggplot(data = df) +
  geom_point(aes(x = RCTD.mean, y = STdeconvolve.mean, color = correlation)) +
  scale_color_viridis(option = "A", direction = -1) +
  theme_classic() +
  # theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(title = "Correlation of predicted proportions of paired cell types",
       x = "RCTD mean spot proportion", y = "STdeconvolve mean spot proportion")
ggsave(filename = "Fig_S_mob-STdev-RCTD-ct_corrs.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

```

```{r}

bottom <- order(lda_vs_sl_theta, decreasing = FALSE)[1:5]

corMtx <- getCorrMtx(m1 = sl_ctTheta, # rows
                           m2 = mob_k38$theta, # columns
                           type = "t")
pairs <- lsatPairs(corMtx)

sl_order <- rownames(corMtx[pairs$rowix, pairs$colsix])
sl_means <- colMeans(sl_ctTheta)[sl_order]

lda_order <- colnames(corMtx[pairs$rowix, pairs$colsix])
lda_means <- colMeans(mob_k38$theta)[lda_order]

df <- data.frame("SL mean" = sl_means,
                 "STdeconvolve mean" = lda_means,
                 "SL ct" = sl_order,
                 "STdeconvolve ct" = lda_order,
                 "correlation" = lda_vs_sl_theta,
                 "rank" = rank(lda_vs_sl_theta))

corMtx[pairs$rowix[bottom], pairs$colsix[bottom]]
df

ggplot(data = df) +
  geom_point(aes(x = SL.mean, y = STdeconvolve.mean, color = correlation)) +
  scale_color_viridis(option = "A", direction = -1) +
  theme_classic() +
  # theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(title = "Correlation of predicted proportions of paired cell types",
       x = "SPOTlight mean spot proportion", y = "STdeconvolve mean spot proportion")
ggsave(filename = "Fig_S_mob-STdev-SL-ct_corrs.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

```

# LDA vs Truth

## k=38

## theta txn clusters

```{r, fig.height=8, fig.width=10}

corMtx <- getCorrMtx(m1 = as.matrix(mobProxyTheta),
                         m2 = mob_k38$theta,
                         type = "t")

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Predicted topics",
                  # ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,11)) # margins for labels
# dev.off()

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor_pairs.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(corMtx)
gplots::heatmap.2(x = corMtx[pairs$rowix, pairs$colsix],
                  Rowv = NA, # to order topics based on pairs and not dendrogram
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Predicted topics",
                  # ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,11))
# dev.off()

diag_cors <- diag(corMtx[pairs$rowix, pairs$colsix])

hist(diag_cors, breaks = 10)
mean(diag_cors)
sd(diag_cors)

median(diag_cors)
range(diag_cors)

lda_vs_txn_theta <- diag_cors

```

# Remove OECs

## SL Before

```{r, fig.height=8, fig.width=10}

corMtx <- getCorrMtx(m1 = as.matrix(mobProxyTheta),
                         m2 = sl_ctTheta,
                         type = "t")

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Predicted topics",
                  # ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,11)) # margins for labels
# dev.off()

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor_pairs.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(corMtx)
gplots::heatmap.2(x = corMtx[pairs$rowix, pairs$colsix],
                  Rowv = NA, # to order topics based on pairs and not dendrogram
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Predicted topics",
                  # ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,11))
# dev.off()

diag_cors <- diag(corMtx[pairs$rowix, pairs$colsix])

hist(diag_cors, breaks = 10)
mean(diag_cors)
sd(diag_cors)

median(diag_cors)
range(diag_cors)

sl_vs_txn_theta <- diag_cors

```

OECs correlate with the Olfactory Nerve Layer

N2 correlates with the Granule Cell Layer.

```{r}

m <- sl_ctTheta[,c("OEC1","OEC2","OEC3","OEC4","OEC5", "N2")]
p <- mobCorpus$mob$pos

# cc <- as.factor(rainbow(ncol(m)))
# names(cc) <- colnames(m)
# 
# vizTopicClusters(theta = m,
#                  pos = p,
#                  clusters = cc,
#                  sharedCol = TRUE,
#                  groups = as.character(mobProxyTheta$`Olfactory Nerve Layer`),
#                  group_cols = c("0" = "lightgray", "1" = "black"),
#                  r = 0.4, # different size piecharts for mOB data sets
#                  lwd = 0.3,
#                  showLegend = TRUE,
#                  plotTitle = NA)

# make OECs the same cluster/level
cc_OECs <- c("#FF0000", "#FF0000", "#FF0000", "#FF0000", "#FF0000", "#FF00FF")
names(cc_OECs) <- c("OEC1","OEC2","OEC3","OEC4","OEC5", "N2")
cc_OECs <- as.factor(cc_OECs)

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc_OECs,
                 sharedCol = TRUE,
                 groups = as.character(mobProxyTheta$`Olfactory Nerve Layer`),
                 group_cols = c("0" = "lightgray", "1" = "black"),
                 r = 0.4, # different size piecharts for mOB data sets
                 lwd = 0.1,
                 showLegend = TRUE,
                 plotTitle = NA,
                 fig_path = fig_path,
                 fig_prefix = "olf_nerve")

cc_OECs <- c("#FF0000", "#FF0000", "#FF0000", "#FF0000", "#FF0000", "#FF00FF")
names(cc_OECs) <- c("OEC1","OEC2","OEC3","OEC4","OEC5", "N2")
cc_OECs <- as.factor(cc_OECs)

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc_OECs,
                 sharedCol = TRUE,
                 groups = as.character(mobProxyTheta$`Granule Cell Layer`),
                 group_cols = c("0" = "lightgray", "1" = "black"),
                 r = 0.4, # different size piecharts for mOB data sets
                 lwd = 0.1,
                 showLegend = TRUE,
                 plotTitle = NA,
                 fig_path = fig_path,
                 fig_prefix = "granule")


```

## SL After

```{r, fig.height=8, fig.width=10}

corMtx <- getCorrMtx(m1 = as.matrix(mobProxyTheta),
                         m2 = sl_countsClean_noOEC$thetaCt,
                         type = "t")

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Predicted topics",
                  # ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,11)) # margins for labels
# dev.off()

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor_pairs.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(corMtx)
gplots::heatmap.2(x = corMtx[pairs$rowix, pairs$colsix],
                  Rowv = NA, # to order topics based on pairs and not dendrogram
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Predicted topics",
                  # ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,11))
# dev.off()

diag_cors <- diag(corMtx[pairs$rowix, pairs$colsix])

hist(diag_cors, breaks = 10)
mean(diag_cors)
sd(diag_cors)

median(diag_cors)
range(diag_cors)

```

N2 shifts to Olfactory Nerve Layer

```{r}

m <- sl_countsClean_noOEC$thetaCt
p <- mobCorpus$mob$pos

# cc <- as.factor(rainbow(ncol(m)))
# names(cc) <- colnames(m)

cc <- c("#FF00FF")
names(cc) <- c("N2")
cc <- as.factor(cc)

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc,
                 sharedCol = TRUE,
                 groups = as.character(mobProxyTheta$`Granule Cell Layer`),
                 group_cols = c("0" = "lightgray", "1" = "black"),
                 r = 0.4, # different size piecharts for mOB data sets
                 lwd = 0.1,
                 showLegend = TRUE,
                 plotTitle = NA,
                 fig_path = fig_path,
                 fig_prefix = "granule_noOEC")

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc,
                 sharedCol = TRUE,
                 groups = as.character(mobProxyTheta$`Olfactory Nerve Layer`),
                 group_cols = c("0" = "lightgray", "1" = "black"),
                 r = 0.4, # different size piecharts for mOB data sets
                 lwd = 0.1,
                 showLegend = TRUE,
                 plotTitle = NA,
                 fig_path = fig_path,
                 fig_prefix = "olf_nerve_noOEC")

```

Is N2 transcriptionally similar to OECs?

## transcriptional similarity

```{r, fig.height=8, fig.width=8}

corMtx <- getCorrMtx(m1 = sl_countsClean$betaCt,
                         m2 = sl_countsClean$betaCt, # gt made from cells in bregma
                         type = "b")

# pdf(file = paste0(fig_path, "Fig1_B-5_FN7_k8_mjrcts_betaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Major cell class",
                  # ylab = "Predicted topics",
                  key.title = NA,
                  cexRow = 0.9,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(7,4))
# dev.off()

```

N2 does cluster with the OECs. So makes sense model would assign this in place of the OECs. It's interesting that it ends up in the Granule Cell Layer otherwise.

And for the actual scRNAseq data:

```{r, fig.height=8, fig.width=8}

corMtx <- getCorrMtx(m1 = mobWtCellProxyBetaRaw[,colnames(sl_countsClean$betaCt)], # reduce 18560 genes to the 6336
                         m2 = mobWtCellProxyBetaRaw[,colnames(sl_countsClean$betaCt)],
                         type = "b")

# pdf(file = paste0(fig_path, "Fig1_B-5_FN7_k8_mjrcts_betaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Major cell class",
                  # ylab = "Predicted topics",
                  key.title = NA,
                  cexRow = 0.9,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(7,4))
# dev.off()

heatmap(x = corMtx)

```

N2 does fall in with OECs here too. Makes sense. SL NMFs fitted with the same scRNAseq data.

It is interesting comparing the auto-correlations of the SL beta (the W) and the actual raw scRNAseq cluster average gene counts.

## RCTD Before

```{r, fig.height=8, fig.width=10}

corMtx <- getCorrMtx(m1 = as.matrix(mobProxyTheta),
                         m2 = as.matrix(RCTD_mob_results_norm),
                         type = "t")

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Predicted topics",
                  # ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,11)) # margins for labels
# dev.off()

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor_pairs.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(corMtx)
gplots::heatmap.2(x = corMtx[pairs$rowix, pairs$colsix],
                  Rowv = NA, # to order topics based on pairs and not dendrogram
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Predicted topics",
                  # ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,11))
# dev.off()

diag_cors <- diag(corMtx[pairs$rowix, pairs$colsix])

hist(diag_cors, breaks = 10)
mean(diag_cors)
sd(diag_cors)

median(diag_cors)
range(diag_cors)

rctd_vs_txn_theta <- diag_cors

```

OECs correspond to Olfactory Nerve Layer as expected.
N2 was not one of the detected 28 cell type clusters.


```{r}

m <- as.matrix(RCTD_mob_results_norm)[,c("OEC1","OEC2","OEC3","OEC4","OEC5")]
p <- mobCorpus$mob$pos

# cc <- as.factor(rainbow(ncol(m)))
# names(cc) <- colnames(m)
# 
# vizTopicClusters(theta = m,
#                  pos = p,
#                  clusters = cc,
#                  sharedCol = TRUE,
#                  groups = as.character(mobProxyTheta$`Olfactory Nerve Layer`),
#                  group_cols = c("0" = "lightgray", "1" = "black"),
#                  r = 0.4, # different size piecharts for mOB data sets
#                  lwd = 0.3,
#                  showLegend = TRUE,
#                  plotTitle = NA)

# make OECs the same cluster/level
cc_OECs <- c("#FF0000", "#FF0000", "#FF0000", "#FF0000", "#FF0000")
names(cc_OECs) <- c("OEC1","OEC2","OEC3","OEC4","OEC5")
cc_OECs <- as.factor(cc_OECs)

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc_OECs,
                 sharedCol = TRUE,
                 groups = as.character(mobProxyTheta$`Olfactory Nerve Layer`),
                 group_cols = c("0" = "lightgray", "1" = "black"),
                 r = 0.4, # different size piecharts for mOB data sets
                 lwd = 0.1,
                 showLegend = TRUE,
                 plotTitle = NA,
                 fig_path = fig_path,
                 fig_prefix = "olf_nerve_rctd")

cc_OECs <- c("#FF0000", "#FF0000", "#FF0000", "#FF0000", "#FF0000")
names(cc_OECs) <- c("OEC1","OEC2","OEC3","OEC4","OEC5")
cc_OECs <- as.factor(cc_OECs)

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc_OECs,
                 sharedCol = TRUE,
                 groups = as.character(mobProxyTheta$`Granule Cell Layer`),
                 group_cols = c("0" = "lightgray", "1" = "black"),
                 r = 0.4, # different size piecharts for mOB data sets
                 lwd = 0.1,
                 showLegend = TRUE,
                 plotTitle = NA,
                 fig_path = fig_path,
                 fig_prefix = "granule_rctd")

```

## RCTD After

```{r, fig.height=8, fig.width=10}

corMtx <- getCorrMtx(m1 = as.matrix(mobProxyTheta),
                         m2 = as.matrix(RCTD_mob_noOEC_results_norm),
                         type = "t")

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Predicted topics",
                  # ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,11)) # margins for labels
# dev.off()

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor_pairs.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(corMtx)
gplots::heatmap.2(x = corMtx[pairs$rowix, pairs$colsix],
                  Rowv = NA, # to order topics based on pairs and not dendrogram
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Predicted topics",
                  # ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 1,
                  # lhei = c(1,3),
                  margins = c(6,11))
# dev.off()

diag_cors <- diag(corMtx[pairs$rowix, pairs$colsix])

hist(diag_cors, breaks = 10)
mean(diag_cors)
sd(diag_cors)

median(diag_cors)
range(diag_cors)

```

N2 strongly correlates with Olfactory Nerve Layer now.

```{r}

m <- as.matrix(RCTD_mob_noOEC_results_norm)
p <- mobCorpus$mob$pos

# cc <- as.factor(rainbow(ncol(m)))
# names(cc) <- colnames(m)

cc <- c("#FF00FF")
names(cc) <- c("N2")
cc <- as.factor(cc)

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc,
                 sharedCol = TRUE,
                 groups = as.character(mobProxyTheta$`Granule Cell Layer`),
                 group_cols = c("0" = "lightgray", "1" = "black"),
                 r = 0.4, # different size piecharts for mOB data sets
                 lwd = 0.1,
                 showLegend = TRUE,
                 plotTitle = NA,
                 fig_path = fig_path,
                 fig_prefix = "granule_rctd_noOEC")

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc,
                 sharedCol = TRUE,
                 groups = as.character(mobProxyTheta$`Olfactory Nerve Layer`),
                 group_cols = c("0" = "lightgray", "1" = "black"),
                 r = 0.4, # different size piecharts for mOB data sets
                 lwd = 0.1,
                 showLegend = TRUE,
                 plotTitle = NA,
                 fig_path = fig_path,
                 fig_prefix = "olf_nerve_rctd_noOEC")

```

## gexp N2

```{r}

# differential expressed marker genes for N2 cluster

n2_markers <- cluster_markers_mob_se_wt_raw[cluster_markers_mob_se_wt_raw$cluster == "N2",]
n2_markers[n2_markers$gene %in% rownames(mob_cpm),]

```

```{r}

# input df for `vizGeneCounts`
marker_genes <- c("Igfbpl1", "Prrg3", "Vcan")

df <- t(mob_cpm)[,marker_genes]
df <- merge(mobCorpus$mob$pos, as.data.frame(df), by=0)
rownames(df) <- df$Row.names

for (gene in marker_genes){
  vizGeneCounts(df = df,
              gene = gene,
              groups = as.character(mobProxyTheta$`Granule Cell Layer`),
              group_cols = c("0" = "white", "1" = "black"),
              size = 7, stroke = 0.5,
              plotTitle = gene,
              showLegend = TRUE)
  ggsave(filename = paste0("N2_gexp_granule_", gene, ".pdf"),
         device = "pdf",
         path = fig_path,
         scale = 1.5,
         width = 5,
         height = 4,
         units = c("in"),
         dpi = 600)
  
  vizGeneCounts(df = df,
              gene = gene,
              groups = as.character(mobProxyTheta$`Olfactory Nerve Layer`),
              group_cols = c("0" = "white", "1" = "black"),
              size = 7, stroke = 0.5,
              plotTitle = gene,
              showLegend = TRUE)
  ggsave(filename = paste0("N2_gexp_olf_nerve_", gene, ".pdf"),
         device = "pdf",
         path = fig_path,
         scale = 1.5,
         width = 5,
         height = 4,
         units = c("in"),
         dpi = 600)
  
}

```

```{r}

for



vizGeneCounts(df = df,
              gene = "Apod",
              groups = NA,
              group_cols = NA,
              size = 7, stroke = 0.01,
              plotTitle = "Apod",
              showLegend = TRUE)
ggsave(filename = "Fig1_F-1_mob_k15_cpm_gexp_Apod.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

vizGeneCounts(df = df,
              gene = "Fabp7",
              groups = NA,
              group_cols = NA,
              size = 7, stroke = 0.01,
              plotTitle = "Fabp7",
              showLegend = TRUE)
ggsave(filename = "Fig1_F-1_mob_k15_cpm_gexp_Fabp7.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

vizGeneCounts(df = df,
              gene = "Igf2",
              groups = NA,
              group_cols = NA,
              size = 7, stroke = 0.01,
              plotTitle = "Igf2",
              showLegend = TRUE)

vizGeneCounts(df = df,
              gene = "Dcn",
              groups = NA,
              group_cols = NA,
              size = 7, stroke = 0.01,
              plotTitle = "Dcn",
              showLegend = TRUE)

```


# Cortex Reference

## SL

```{r}

# L5.IT not present at all in SL so drop
sl_cortex_ctTheta <- sl_countsClean_cortexRef$thetaCt[,which(!colnames(sl_countsClean_cortexRef$thetaCt) %in% c("L5.IT"))]

# 22 cell types now

```

```{r, fig.height=8, fig.width=10}

corMtx <- getCorrMtx(m1 = sl_cortex_ctTheta,
                         m2 = sl_ctTheta,
                         type = "t")

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "MOB",
                  ylab = "Cortex",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,8)) # margins for labels
# dev.off()

# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(corMtx)
# new mtx where rows and cols in order in which they are paired
corMtx_paired <- corMtx[pairs$rowix, pairs$colsix]

# get dendro clusters for the mob cts based on the gexp similarity
mob_sig_marker_genes <- unique(cluster_markers_mob_se_wt_raw[cluster_markers_mob_se_wt_raw$p_val_adj < 0.05,]$gene)
beta <- mobWtCellProxyBetaRaw[colnames(corMtx_paired), mob_sig_marker_genes]
# beta <- sl_countsClean$betaCt[colnames(corMtx_paired),]
d_ <- as.dist(1-cor(t(beta)))
hc_ <- stats::hclust(d_, method = "ward.D")

pdf(file = paste0(fig_path, "Fig_S6_SL-mob-vs-cortex-corr-heatmap.pdf"))
gplots::heatmap.2(x = corMtx_paired[hc_$order,],
                  Rowv = NA, # to order topics based on pairs and not dendrogram
                  Colv = as.dendrogram(hc_),
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "MOB",
                  ylab = "Cortex",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 1,
                  # lhei = c(1,3),
                  margins = c(6,8))
dev.off()

diag_cors <- diag(corMtx_paired)

hist(diag_cors, breaks = 10)
mean(diag_cors)
sd(diag_cors)

median(diag_cors)
range(diag_cors)

pdf(file = paste0(fig_path, "Fig_S6_SL-mob-vs-cortex-corr-boxplot.pdf"),   # The directory you want to save the file in
    width = 5, # The width of the plot in inches
    height = 3) # The height of the plot in inches
par( mfrow = c(1,3) )
boxplot(diag_cors,
        medcol = "black", # medium line color
        medlty = 1, # line type
        medlwd = 1, # line width
        boxfill = "white", # color fill of box
        boxcol = "black", # outline color of box
        boxlty = 1, # box line type
        boxlwd = 1, # width of box line
        whisklty = 1, # whisker type
        whisklwd = 1,
        staplelwd = 0, # the ends of the whiskers
        outpch = 1, # type of outlies
        outcex = 0,
        main = "",
        cex.main = 1,
        xlab = "",
        ylab = "Correlation") # size of outliers
stripchart(diag_cors, vertical = TRUE,
    method = "jitter", add = TRUE, pch = 20, col = 'black')
dev.off()

```

the mob cts have been clustered by gexp (beta) and the paired cortex cts also paired up. This correlation based on theta. What about the gexp correlation between mob and cortex cts?


Above is wrt theta, so predicted positions. VLMC is being placed where the OECs are assigned (the Olfactory Nerve Layer)

```{r}

m <- sl_cortex_ctTheta
p <- mobCorpus$mob$pos

# cc <- as.factor(rainbow(ncol(m)))
# names(cc) <- colnames(m)

cc <- c("#FF00FF")
names(cc) <- c("VLMC")
cc <- as.factor(cc)

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc,
                 sharedCol = TRUE,
                 groups = as.character(mobProxyTheta$`Olfactory Nerve Layer`),
                 group_cols = c("0" = "white", "1" = "black"),
                 r = 0.4, # different size piecharts for mOB data sets
                 lwd = 0.3,
                 showLegend = TRUE,
                 plotTitle = NA)

```

```{r, fig.height=8, fig.width=10}

corMtx <- getCorrMtx(m1 = sl_countsClean_cortexRef$betaCt,
                         m2 = sl_countsClean$betaCt,
                         type = "b")

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "MOB",
                  ylab = "Cortex",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,8)) # margins for labels
# dev.off()

pdf(file = paste0(fig_path, "Fig_S6_rctd-mob-vs-cortex-corr-heatmap.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(corMtx)
gplots::heatmap.2(x = corMtx[pairs$rowix, pairs$colsix],
                  Rowv = NA, # to order topics based on pairs and not dendrogram
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "MOB",
                  ylab = "Cortex",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 1,
                  # lhei = c(1,3),
                  margins = c(6,8))
dev.off()

diag_cors <- diag(corMtx[pairs$rowix, pairs$colsix])

hist(diag_cors, breaks = 10)
mean(diag_cors)
sd(diag_cors)

median(diag_cors)
range(diag_cors)

pdf(file = paste0(fig_path, "Fig_S6_rctd-mob-vs-cortex-corr-boxplot.pdf"),   # The directory you want to save the file in
    width = 5, # The width of the plot in inches
    height = 3) # The height of the plot in inches
par( mfrow = c(1,3) )
boxplot(diag_cors,
        medcol = "black", # medium line color
        medlty = 1, # line type
        medlwd = 1, # line width
        boxfill = "white", # color fill of box
        boxcol = "black", # outline color of box
        boxlty = 1, # box line type
        boxlwd = 1, # width of box line
        whisklty = 1, # whisker type
        whisklwd = 1,
        staplelwd = 0, # the ends of the whiskers
        outpch = 1, # type of outlies
        outcex = 0,
        main = "",
        cex.main = 1,
        xlab = "",
        ylab = "Correlation") # size of outliers
stripchart(diag_cors, vertical = TRUE, 
    method = "jitter", add = TRUE, pch = 20, col = 'black')
dev.off()

```

Transcriptionally, VLMC actually pairs with Mes1 the best. Some weak correlation with OECs, particularly 1 and 2, which are most abundant of the OECs in the original mob SL prediction.

```{r}

m <- sl_ctTheta
p <- mobCorpus$mob$pos

# cc <- as.factor(rainbow(ncol(m)))
# names(cc) <- colnames(m)

cc <- c("#FF00FF")
names(cc) <- c("Mes1")
cc <- as.factor(cc)

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc,
                 sharedCol = TRUE,
                 groups = as.character(mobProxyTheta$`Olfactory Nerve Layer`),
                 group_cols = c("0" = "white", "1" = "black"),
                 r = 0.4, # different size piecharts for mOB data sets
                 lwd = 0.3,
                 showLegend = TRUE,
                 plotTitle = NA)

```

Mes1 slightly correlates with VLMC in terms of spatial proportions but only because VLMC touches it at the top.


Lets take a look at VLMC and how it relates transcriptionally to the mob clusters using the scRNAseq gene counts

### gexp

```{r, fig.height=8, fig.width=8}

# in terms of cluster genes for the mob and cortex cts
corMtx <- getCorrMtx(m1 = cortexProxyBetaRaw[,colnames(sl_countsClean_cortexRef$betaCt)],
                     m2 = mobWtCellProxyBetaRaw[,colnames(sl_countsClean$betaCt)],
                     type = "b")

# pdf(file = paste0(fig_path, "Fig1_B-4_FN7_k8_mjrcts_thetaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "MOB",
                  ylab = "Cortex",
                  key.title = NA,
                  cexRow = 0.8,
                  cexCol = 0.8,
                  # lhei = c(1,3),
                  margins = c(7,4))
# dev.off()

# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(corMtx)
corMtx_paired <- corMtx[pairs$rowix, pairs$colsix]

# get dendro clusters for the mob cts based on the gexp similarity
mob_sig_marker_genes <- unique(cluster_markers_mob_se_wt_raw[cluster_markers_mob_se_wt_raw$p_val_adj < 0.05,]$gene)
beta <- mobWtCellProxyBetaRaw[colnames(corMtx_paired), mob_sig_marker_genes]
# beta <- sl_countsClean$betaCt[colnames(corMtx_paired),]
d_ <- as.dist(1-cor(t(beta)))
hc_ <- stats::hclust(d_, method = "ward.D")

pdf(file = paste0(fig_path, "Fig_S6_mob-vs-cortex-gexp-heatmap.pdf"))
gplots::heatmap.2(x = corMtx_paired[hc_$order,],
                  Rowv = NA, # to order topics based on pairs and not dendrogram
                  Colv = as.dendrogram(hc_),
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "MOB",
                  ylab = "Cortex",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 1,
                  # lhei = c(1,3),
                  margins = c(6,8))
dev.off()

diag_cors <- diag(corMtx_paired)

hist(diag_cors, breaks = 10)
mean(diag_cors)
sd(diag_cors)

median(diag_cors)
range(diag_cors)

pdf(file = paste0(fig_path, "Fig_S6_mob-vs-cortex-gexp-boxplot.pdf"),   # The directory you want to save the file in
    width = 5, # The width of the plot in inches
    height = 3) # The height of the plot in inches
par( mfrow = c(1,3) )
boxplot(diag_cors,
        medcol = "black", # medium line color
        medlty = 1, # line type
        medlwd = 1, # line width
        boxfill = "white", # color fill of box
        boxcol = "black", # outline color of box
        boxlty = 1, # box line type
        boxlwd = 1, # width of box line
        whisklty = 1, # whisker type
        whisklwd = 1,
        staplelwd = 0, # the ends of the whiskers
        outpch = 1, # type of outlies
        outcex = 0,
        main = "",
        cex.main = 1,
        xlab = "",
        ylab = "Correlation") # size of outliers
stripchart(diag_cors, vertical = TRUE,
    method = "jitter", add = TRUE, pch = 20, col = 'black')
dev.off()

```

Transcriptionally, VLMC highly similar to Mes1

## RCTD

```{r, fig.height=8, fig.width=8}

corMtx <- getCorrMtx(m1 = as.matrix(RCTD_cortexRef_results_norm),
                     m2 = as.matrix(RCTD_mob_results_norm),
                     type = "t")

# pdf(file = paste0(fig_path, "Fig1_B-4_FN7_k8_mjrcts_thetaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "MOB",
                  ylab = "Cortex",
                  key.title = NA,
                  cexRow = 0.8,
                  cexCol = 0.8,
                  # lhei = c(1,3),
                  margins = c(7,4))
# dev.off()

# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(corMtx)
# new mtx where rows and cols in order in which they are paired
corMtx_paired <- corMtx[pairs$rowix, pairs$colsix]

# get dendro clusters for the mob cts based on the gexp similarity
mob_sig_marker_genes <- unique(cluster_markers_mob_se_wt_raw[cluster_markers_mob_se_wt_raw$p_val_adj < 0.05,]$gene)
beta <- mobWtCellProxyBetaRaw[colnames(corMtx_paired), mob_sig_marker_genes]
# beta <- sl_countsClean$betaCt[colnames(corMtx_paired),]
d_ <- as.dist(1-cor(t(beta)))
hc_ <- stats::hclust(d_, method = "ward.D")

pdf(file = paste0(fig_path, "Fig_S6_RCTD-mob-vs-cortex-corr-heatmap.pdf"))
gplots::heatmap.2(x = corMtx_paired[hc_$order,],
                  Rowv = NA, # to order topics based on pairs and not dendrogram
                  Colv = as.dendrogram(hc_),
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  xlab = "MOB",
                  ylab = "Cortex",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 1,
                  # lhei = c(1,3),
                  margins = c(6,8))
dev.off()

diag_cors <- diag(corMtx_paired)

hist(diag_cors, breaks = 10)
mean(diag_cors)
sd(diag_cors)

median(diag_cors)
range(diag_cors)

pdf(file = paste0(fig_path, "Fig_S6_RCTD-mob-vs-cortex-corr-boxplot.pdf"),   # The directory you want to save the file in
    width = 5, # The width of the plot in inches
    height = 3) # The height of the plot in inches
par( mfrow = c(1,3) )
boxplot(diag_cors,
        medcol = "black", # medium line color
        medlty = 1, # line type
        medlwd = 1, # line width
        boxfill = "white", # color fill of box
        boxcol = "black", # outline color of box
        boxlty = 1, # box line type
        boxlwd = 1, # width of box line
        whisklty = 1, # whisker type
        whisklwd = 1,
        staplelwd = 0, # the ends of the whiskers
        outpch = 1, # type of outlies
        outcex = 0,
        main = "",
        cex.main = 1,
        xlab = "",
        ylab = "Correlation") # size of outliers
stripchart(diag_cors, vertical = TRUE,
    method = "jitter", add = TRUE, pch = 20, col = 'black')
dev.off()

```

```{r}

bottom <- order(diag_cors, decreasing = FALSE)[1:10]

corMtx <- getCorrMtx(m1 = as.matrix(RCTD_cortexRef_results_norm),
                     m2 = as.matrix(RCTD_mob_results_norm),
                     type = "t")
pairs <- lsatPairs(corMtx)

rctd_order <- colnames(corMtx[pairs$rowix, pairs$colsix])
rctd_means <- colMeans(as.matrix(RCTD_mob_results_norm))[rctd_order]

cortex_order <- rownames(corMtx[pairs$rowix, pairs$colsix])
cortex_means <- colMeans(as.matrix(RCTD_cortexRef_results_norm))[cortex_order]

df <- data.frame("RCTD mean" = rctd_means,
                 "Cortex mean" = cortex_means,
                 "RCTD ct" = rctd_order,
                 "Cortex ct" = cortex_order,
                 "correlation" = diag_cors,
                 "rank" = rank(diag_cors))

corMtx[pairs$rowix[bottom], pairs$colsix[bottom]]
df

ggplot(data = df) +
  geom_point(aes(x = RCTD.mean, y = Cortex.mean, color = correlation)) +
  scale_color_viridis(option = "A", direction = -1) +
  theme_classic() +
  # theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(title = "RCTD Correlation of predicted proportions of paired cell types",
       x = "MOB ref mean spot proportion", y = "Cortex ref mean spot proportion")
# ggsave(filename = "Fig_S_mob-RCTD-SL-ct_corrs.pdf",
#        device = "pdf",
#        path = fig_path,
#        scale = 1.5,
#        width = 5,
#        height = 4,
#        units = c("in"),
#        dpi = 600)

```

VLMC also correlates in terms of proportions with Mes1 and OECs.

Again, likely due to its transcriptional similarity, at least to Mes1. Partial correlation with OECs, but interesting not very strong.

Looking at the scRNAseq data, transcriptionally, the OECs don't correlate that strongly with anything except maybe VLMC and the Astros, which may hve been assigned elsewhere.

```{r}

m <- as.matrix(RCTD_cortexRef_results_norm)
p <- mobCorpus$mob$pos

# cc <- as.factor(rainbow(ncol(m)))
# names(cc) <- colnames(m)

cc <- c("#FF00FF")
names(cc) <- c("VLMC")
cc <- as.factor(cc)

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc,
                 sharedCol = TRUE,
                 groups = as.character(mobProxyTheta$`Olfactory Nerve Layer`),
                 group_cols = c("0" = "lightgray", "1" = "black"),
                 r = 0.4, # different size piecharts for mOB data sets
                 lwd = 0.1,
                 showLegend = TRUE,
                 plotTitle = NA,
                 fig_path = fig_path,
                 fig_prefix = "olf_nerve_rctd_cortex")

```

# Compare Methods vs Truth

```{r}

txn_theta <- data.frame("SPOTlight" = as.vector(sl_vs_txn_theta),
                        "RCTD" = as.vector(rctd_vs_txn_theta),
                        "STdeconvolve" = as.vector(lda_vs_txn_theta))

ggplot(data = reshape2::melt(txn_theta),
       aes(x = variable, y = value, fill = variable)) +
  geom_boxplot(outlier.shape=NA) +
  geom_point(position = position_jitterdodge(), alpha = 1, size = 1) + 
  labs(title = "MOB predicted proportions versus txn clusters",
       x = "Method", y = "Correlation") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
ggsave(filename = "Fig_2_mob-methods-corr-txn-clusters.pdf",
       device = "pdf",
       path = fig_path,
       scale = 1.5,
       width = 5,
       height = 4,
       units = c("in"),
       dpi = 600)

```

# GSEA for betas

## viz mob_k15 topics and txn clusts

```{r}

m <- mob_k15$theta # theta for the k=15 fitted LDA for mob data set
p <- mobCorpus$mob$pos # the positions of filtered spots for mob data set

vizAllTopics(theta = m,
             pos = p,
             topicOrder=seq(ncol(m)),
             topicCols=rainbow(ncol(m)),
             groups = NA,
             group_cols = NA,
             r = 0.4, # different size piecharts for mOB data sets
             lwd = 0.01,
             showLegend = TRUE,
             plotTitle = NA)

```

```{r, mob-qc, fig.width=8, fig.height=3}

# data("mOB")
# 
# # Remove poor datasets and genes
# mob_counts <- MERINGUE::cleanCounts(counts = mOB$counts, # from mOB data set
#                       min.reads = 100, 
#                       min.lib.size = 100, 
#                       plot=TRUE,
#                       verbose=TRUE)
# mob_pos <- mOB$pos[colnames(mob_counts),]

# CPM normalize
mob_cpm <- MERINGUE::normalizeCounts(counts = t(countsClean), 
                       log=FALSE,
                       verbose=TRUE)

```

```{r, fig.height=4, fig.width=8}

set.seed(333)

# Dimensionality reduction by PCA on log10 CPM expression values
pcs.info <- prcomp(t(log10(as.matrix(mob_cpm)+1)), center=TRUE)
nPcs <- 5
pcs <- pcs.info$x[,1:nPcs]

# 2D embedding by tSNE
emb <- Rtsne::Rtsne(pcs,
             is_distance=FALSE,
             perplexity=30,
             num_threads=1,
             verbose=FALSE)$Y
rownames(emb) <- rownames(pcs)

# Graph-based cluster detection
k <- 30
com <- getClusters(pcs, k, weight=TRUE)

# Manually annotate identified clusters with cell-types
mob_annot <- as.character(com); names(mob_annot) <- names(com)
mob_annot[com==4] <- '1: Granule Cell Layer'
mob_annot[com==1] <- '2: Mitral Cell Layer'
mob_annot[com==3] <- '3: Outer Plexiform Layer'
mob_annot[com==2] <- '4: Glomerular Layer'
mob_annot[com==5] <- '5: Olfactory Nerve Layer'
mob_annot <- as.factor(mob_annot)

# Plot
pdf(file = paste0(fig_path, "Fig_S_mob-txn.pdf"),   # The directory you want to save the file in
    width = 7, # The width of the plot in inches
    height = 6) # The height of the plot in inches
par(mfrow=c(2,2), mar=rep(1,4))
plotEmbedding(emb, groups=mob_annot, 
              show.legend=TRUE, xlab=NA, ylab=NA,
              verbose=FALSE)
plotEmbedding(mobCorpus$mob$pos, groups=mob_annot, 
              cex=1, xlab=NA, ylab=NA,
              verbose=FALSE)
dev.off()

```

```{r, fig.height=8, fig.width=10}

corMtx <- getCorrMtx(m1 = as.matrix(mobProxyTheta),
                         m2 = mob_k15$theta,
                         type = "t")

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor.pdf"))
gplots::heatmap.2(x = corMtx,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Predicted topics",
                  # ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,11)) # margins for labels
# dev.off()

# pdf(file = paste0(fig_path, "Fig1_E-3_mob_k15_thetaCor_pairs.pdf"))
# pair up best matching cell class to a predicted topic
pairs <- lsatPairs(corMtx)
gplots::heatmap.2(x = corMtx[pairs$rowix, pairs$colsix],
                  Rowv = NA, # to order topics based on pairs and not dendrogram
                  Colv = NA,
                  density.info = "none",
                  trace = "none",
                  col = correlation_palette,
                  breaks = correlation_breaks,
                  key.xlab = "Correlation",
                  # xlab = "Predicted topics",
                  # ylab = "Cell layer",
                  key.title = NA,
                  cexRow = 1,
                  cexCol = 0.9,
                  # lhei = c(1,3),
                  margins = c(6,11))
# dev.off()

```

```{r}

m <- mob_k15$theta
p <- mobCorpus$mob$pos

# cc <- as.factor(rainbow(ncol(m)))
# names(cc) <- colnames(m)

cc <- rainbow(ncol(m))
names(cc) <- colnames(m)
cc <- as.factor(cc)

vizTopicClusters(theta = m,
                 pos = p,
                 clusters = cc,
                 sharedCol = TRUE,
                 # groups = as.character(mobProxyTheta$`Outer Plexiform Layer`),
                 # group_cols = c("0" = "white", "1" = "black"),
                 r = 0.4, # different size piecharts for mOB data sets
                 lwd = 0.3,
                 showLegend = TRUE,
                 plotTitle = NA)

# vizTopicClusters(theta = m,
#                  pos = p,
#                  clusters = cc,
#                  sharedCol = TRUE,
#                  # groups = as.character(mobProxyTheta$`Mitral Cell Layer`),
#                  # group_cols = c("0" = "white", "1" = "black"),
#                  r = 0.4, # different size piecharts for mOB data sets
#                  lwd = 0.3,
#                  showLegend = TRUE,
#                  plotTitle = NA)

```

## C8 cts

```{r}

# library(msigdbr)
# library(liger)

# create the term - gene list for GSEA C8 Cell type signature gene sets

# all_gene_sets <- msigdbr(species = "Mus musculus")

gene_set <- all_gene_sets %>%
                dplyr::filter(gs_cat == "C8") %>%
                as.data.frame()

# `gs_name` are the cell type labels and the associated genes are the `gene_symbol`
C8_gsea_ct_gene_list <- lapply(unique(gene_set$gs_name), function(ct){
  genes <- gene_set[gene_set$gs_name == ct,]$gene_symbol
  genes
})
names(C8_gsea_ct_gene_list) <- unique(gene_set$gs_name)

```

```{r}

# genes and beta values for a topic
vals <- sort(mob_k15$beta[1,], decreasing=TRUE)
head(vals)

# C8 Cell type signature annotations
gsea.results <- iterative.bulk.gsea(values=vals, set.list=C8_gsea_ct_gene_list, rank=TRUE)
head(gsea.results[order(gsea.results$p.val),], n = 30)

# significant cell type signatures
gsea.sig.up.C8 <- gsea.results[gsea.results$q.val < 0.1 & gsea.results$sscore > 0 & gsea.results$edge > 0,]
head(gsea.sig.up.C8, n = 30)

# viz significant hits
# top <- 1
# gsea(values=vals, geneset=go.env[[rownames(gsea.sig.up.C8)[top]]], plot=TRUE, rank=TRUE)

```

```{r}

# use the index of the C8 ct ID/term to pull out the genes and then plot the combined gene counts
# for the given C8 ct ID term ossociated genes
i <- "GAO_LARGE_INTESTINE_24W_C1_DCLK1POS_PROGENITOR"

# get genes in gene count mat corresponding to the GO term
mat <- t(countsClean)[intersect(C8_gsea_ct_gene_list[[i]], rownames(t(countsClean))),]

# add up counts of all the associated genes for each spot. Scale values
gexp <- scale(colSums(mat))[,1]

plotEmbedding(emb, col=gexp, main=i)
plotEmbedding(mobCorpus$mob$pos, col=gexp, cex = 2)

```

## C2 CP canonical pathways

```{r}

# library(msigdbr)
# library(liger)

# create the term - gene list for GSEA C8 Cell type signature gene sets

# all_gene_sets <- msigdbr(species = "Mus musculus")

gene_set <- all_gene_sets %>%
                dplyr::filter(gs_cat == "C2") %>%
                dplyr::filter(gs_subcat == "CP") %>%
                as.data.frame()

# `gs_name` are the cell type labels and the associated genes are the `gene_symbol`
C2_gsea_gene_list <- lapply(unique(gene_set$gs_name), function(ct){
  genes <- gene_set[gene_set$gs_name == ct,]$gene_symbol
  genes
})
names(C2_gsea_gene_list) <- unique(gene_set$gs_name)

```

```{r}

# genes and beta values for a topic
vals <- sort(mob_k15$beta[1,], decreasing=TRUE)
head(vals)

# C8 Cell type signature annotations
gsea.results <- iterative.bulk.gsea(values=vals, set.list=C2_gsea_gene_list, rank=TRUE)
head(gsea.results[order(gsea.results$p.val),], n = 30)

# significant cell type signatures
gsea.sig.up.C2 <- gsea.results[gsea.results$q.val < 0.1 & gsea.results$sscore > 0 & gsea.results$edge > 0,]
head(gsea.sig.up.C2, n = 30)

# viz significant hits
# top <- 1
# gsea(values=vals, geneset=go.env[[rownames(gsea.sig.up.C8)[top]]], plot=TRUE, rank=TRUE)

```

```{r}

# use the index of the C8 ct ID/term to pull out the genes and then plot the combined gene counts
# for the given C8 ct ID term ossociated genes
i <- "NABA_MATRISOME_ASSOCIATED"

# get genes in gene count mat corresponding to the GO term
mat <- t(countsClean)[intersect(C2_gsea_gene_list[[i]], rownames(t(countsClean))),]

# add up counts of all the associated genes for each spot. Scale values
gexp <- scale(colSums(mat))[,1]

plotEmbedding(emb, col=gexp, main=i)
plotEmbedding(mobCorpus$mob$pos, col=gexp, cex = 2)

```

## C5 GO

```{r}

# library(msigdbr)
# library(liger)

# create the term - gene list for GSEA C8 Cell type signature gene sets

# all_gene_sets <- msigdbr(species = "Mus musculus")

gene_set <- all_gene_sets %>%
                dplyr::filter(gs_cat == "C5") %>%
                dplyr::filter(gs_subcat == "GO:BP") %>%
                as.data.frame()

# `gs_name` are the cell type labels and the associated genes are the `gene_symbol`
C5_gsea_gene_list <- lapply(unique(gene_set$gs_name), function(ct){
  genes <- gene_set[gene_set$gs_name == ct,]$gene_symbol
  genes
})
names(C5_gsea_gene_list) <- unique(gene_set$gs_name)

```

```{r}

# genes and beta values for a topic
vals <- sort(mob_k15$beta[1,], decreasing=TRUE)
head(vals)

# C8 Cell type signature annotations
gsea.results <- iterative.bulk.gsea(values=vals, set.list=C5_gsea_gene_list, rank=TRUE)
head(gsea.results[order(gsea.results$p.val),], n = 30)

# significant cell type signatures
gsea.sig.up.C5 <- gsea.results[gsea.results$q.val < 0.1 & gsea.results$sscore > 0 & gsea.results$edge > 0,]
head(gsea.sig.up.C5, n = 30)

# viz significant hits
# top <- 1
# gsea(values=vals, geneset=go.env[[rownames(gsea.sig.up.C8)[top]]], plot=TRUE, rank=TRUE)

```

```{r}

# use the index of the C8 ct ID/term to pull out the genes and then plot the combined gene counts
# for the given C8 ct ID term ossociated genes
i <- "GO_CELL_CYCLE_PROCESS"

# get genes in gene count mat corresponding to the GO term
mat <- t(countsClean)[intersect(C5_gsea_gene_list[[i]], rownames(t(countsClean))),]

# add up counts of all the associated genes for each spot. Scale values
gexp <- scale(colSums(mat))[,1]

plotEmbedding(emb, col=gexp, main=i)
plotEmbedding(mobCorpus$mob$pos, col=gexp, cex = 2)

```


